---
title: "Covariate analysis of CMC RNASeq data reprocessed with Star-HTSeq pipeline (ACC with TMM normlaisation)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(knitr)
library(knit2synapse)
library(synapseClient)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./CMC_ACC_StarAligned_TMM.Rmd",
                                 parentId ="syn5569976",
                                 entityName = 'ACC - STAR Reprocessed - with TMM')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file

## Load required libraries
library(CovariateAnalysis) # get it from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(reshape2)
library(limma)
library(Biobase)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn5569976';
activityName = 'Covariate adjustments';
activityDescription = 'Covariate analysis of star aligned CMC reprocessed RNASeq data with TMM normalisation (ACC)';

thisFileName <- 'CMC_ACC_StarAligned_TMM.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='CMC')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download reprocessed counts (ACC)
COUNT_ID <- 'syn7217917';
ALL_USED_IDs <- COUNT_ID
COUNT <- read.table(synGet(COUNT_ID)@filePath, header=T, sep='\t', check.names = F, nrows = 56632) %>%
  dplyr::select(-chr, -start, -end, -strand, -length)
rownames(COUNT) = COUNT$geneID
COUNT$geneID = NULL
COUNT = data.matrix(COUNT)

# Get merged metadata
METADATA_ID <- 'syn2299154'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_ID
METADATA <- read.csv(synGet(METADATA_ID)@filePath, fill = T, header = F)

## Metadata with specific information for RNASeq processing
metadataVarOfInterest = c("Individual ID", "Institution", "Gender", "Ethnicity", "Age of Death", "PMI (in hours)", 
                          "pH", "Dx", "CDR", "CauseOfDeath", "ACC_RNA_isolation: Exclude?", "ACC_RNA_isolation: Sample RNA ID",
                          "ACC_RNA_isolation: RNA Isolation Batch", "ACC_RNA_isolation: RIN", "ACC_RNA_report: Exclude?",
                          "ACC_RNA_report: Library Batch")

# Filter metadata 
ind = as.character(METADATA[1,] %>% unlist) %in% metadataVarOfInterest
METADATA = METADATA[,ind]
colnames(METADATA) = gsub('[^[:alnum:]]','_',as.character(METADATA[1,] %>% unlist))

METADATA = METADATA %>%
  dplyr::select(Individual_ID, Institution, Gender, Ethnicity, Age_of_Death, PMI__in_hours_, pH, Dx, CDR, CauseOfDeath,
                ACC_RNA_isolation__Exclude_, ACC_RNA_isolation__Sample_RNA_ID, ACC_RNA_isolation__RNA_Isolation_Batch,
                ACC_RNA_isolation__RIN, ACC_RNA_report__Exclude_, ACC_RNA_report__Library_Batch) %>%
  dplyr::rename(PMI = PMI__in_hours_, 
                IsolationExclude = ACC_RNA_isolation__Exclude_, 
                SampleID = ACC_RNA_isolation__Sample_RNA_ID, 
                IsolationBatch = ACC_RNA_isolation__RNA_Isolation_Batch,
                RIN = ACC_RNA_isolation__RIN, 
                ReportExclude = ACC_RNA_report__Exclude_, 
                LibraryBatch = ACC_RNA_report__Library_Batch) %>%
  dplyr::filter(SampleID %in% colnames(COUNT)) %>%
  droplevels() %>%
  dplyr::mutate(LibraryBatch = gsub('[^[:alnum:]]','_',LibraryBatch),
                LibraryBatch = as.factor(LibraryBatch),
                IsolationBatch = as.factor(IsolationBatch))
```

### Data preprocessing
```{r preprocess.data, echo=TRUE}
# Remove samples marked as exclude and samples with no Gender, Ethnicity, PMI, and RIN
METADATA <- METADATA %>%
  filter(SampleID %in% colnames(COUNT)) %>%
  filter(IsolationExclude == 0, ReportExclude == 0) %>%
  filter(!is.na(PMI)) %>%
  filter(!is.na(Gender)) %>%
  filter(!is.na(Ethnicity)) %>%
  filter(!is.na(RIN)) %>%
  filter(!(SampleID %in% c('MSSM_RNA_ACC_250', 'MSSM_RNA_ACC_170'))) %>%
  as.data.frame()
```

```{r preprocess.data1, results='asis'}
# Match covariates to expression data
indToRetain = intersect(METADATA$SampleID, colnames(COUNT))
removedIDs = setdiff(colnames(COUNT), METADATA$SampleID)

COUNT = COUNT[,indToRetain]

rownames(METADATA) = METADATA$SampleID
METADATA = METADATA[indToRetain,]

METADATA$Dx %>% 
  table() %>% 
  as.data.frame %>% 
  setnames(c('Dx','Total.Number.of.Samples')) %>% 
  kable()
```
Following sample are removed `r paste(removedIDs, collapse = ',')`

### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering}
FactorCovariates <- c("Institution", "Gender", "Ethnicity", "CDR", "CauseOfDeath", "IsolationBatch", "LibraryBatch", "Dx")
ContCovariates <- c("Age_of_Death", "PMI", "pH", "RIN")

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
COVARIATES <- data.frame(lapply(COVARIATES,function(x){x <- sapply(x,function(y){str_replace_all(as.character(y),'\\+','')})}))
rownames(COVARIATES) <- METADATA$SampleID

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)

# Add in RIN^2 values
COVARIATES$RIN2 = COVARIATES$RIN^2
ContCovariates = c(ContCovariates, 'RIN2')
```
Correlation/association between covariates at an FDR <= 0.1
```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.5, row.width=0.1)
```
Explore metatdata
```{r dataExplore, fig.width = 10, fig.height = 10}
my.theme <- theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1))

# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = Dx, y = RIN)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme

# Age of Death
p[[2]] = ggplot(COVARIATES, aes(x = Dx, y = Age_of_Death)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AgeOfDeath') + my.theme

# PMI
p[[3]] = ggplot(COVARIATES, aes(x = Dx, y = PMI)) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PMI (in hours)') + my.theme

# pH
p[[4]] = ggplot(COVARIATES, aes(x = Dx, y = pH)) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('pH') + my.theme

multiplot(plotlist = p, cols = 2)
```
### CPM Normalisation
Preprocess counts matrix and metadata. Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of samples per Dx.
```{r cpmnormalisation}
genesToAnalyze = dlply(METADATA, .(Dx), .fun = function(mtd, count){
  processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$SampleID %>% droplevels()],
                                                   MIN_GENE_CPM=1, 
                                                   MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
  processed.counts$filteredExprMatrix$genes
}, COUNT)
genesToAnalyze = unlist(genesToAnalyze) %>% unique()

PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, ], MIN_GENE_CPM=0, MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)
```{r initial.voom.normalisation}
# TMM normalisation
TMM.GENE_EXPRESSION = calcNormFactors(PROCESSED_COUNTS$filteredExprMatrix)

# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(TMM.GENE_EXPRESSION, design=NULL, plot=T)
```
Coexpression of genes 
```{r coexp1, cache=FALSE}
cr = cor(t(VOOM.GENE_EXPRESSION$E))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```
### Outlier analysis
Clustering of voom adjusted data (with NULL design)
```{r decompse.normalise.data, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Dx, shape=Institution, size=Age_of_Death))
p <- p + theme_bw() + theme(legend.position="top") + scale_shape_manual(values = 1:9)
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=0)
p

# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c('Dx','Institution')])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(VOOM.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp[,c('Dx','Institution')]))
```

```{r temp, include = F}
dev.off()
gc()
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, 
                                                 COVARIATES,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```
Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative design)
Since many covariates are correlated, re-normalising COUNTS with an iterative design matrix. Here "Dx" is chosen as the primary variable of interest
```{r iterativeNorm}
# Primary variable of interest
postAdjustCovars = 'RIN';#'Dx'

# Assign residual covariates
residualSigCovars = preAdjustedSigCovars
covariatesEffects = residualSigCovars$Effects.significantCovars[setdiff(residualSigCovars$significantCovars, postAdjustCovars)]
postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects))) %>% unique()

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 100){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
  
  # Estimate voom weights
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(TMM.GENE_EXPRESSION, design=DM1$design, plot=F)
  
  # Fit linear model using new weights and new design
  VOOM.ADJUSTED.FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), c(postAdjustCovars))
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                COVARIATES[, residCovars, drop=F], 
                                                'adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  covariatesEffects = residualSigCovars$Effects.significantCovars[setdiff(residualSigCovars$significantCovars, 
                                                                          c(postAdjustCovars))]
  
  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}

# Add primary variable to the list (in case not chosen by iterative method)
postAdjustCovars = unique(c('Dx', postAdjustCovars))

writeLines(paste('Using following covariates in the model:',
                 paste(postAdjustCovars, collapse=', '),
                 'as fixed effects'))

# Post adjusted design matrix
DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]

# Estimate voom weights
VOOM.ADJUSTED.GENE_EXPRESSION = voom(TMM.GENE_EXPRESSION, design=DM1$design, plot=F)

# Fit linear model using new weights and new design
VOOM.ADJUSTED.FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION)

# Residuals after normalisation
RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)

# Residual covariates to choose from
residCovars <- setdiff(c(FactorCovariates,ContCovariates), c(postAdjustCovars))

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                              COVARIATES[, residCovars, drop=F], 
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)

modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`
### Sanity check
```{r residualSigCovars.manual, fig.width=20, fig.height=12}
residualSigCovars[["PC_res"]][[2]]$plotData
```
### Residual calculation
Calculate weighted residuals and add back "Dx" to the residuals
```{r varsToAddBack}
# Add variable of interest back to the residuals
varsToAddIn = grep("Dx", colnames(DM1$design), value = T)
RESIDUAL.GENE_EXPRESSION = RESIDUAL.GENE_EXPRESSION + 
  VOOM.ADJUSTED.FIT$coefficients[,varsToAddIn] %*% t(DM1$design[,varsToAddIn])
```
Coexpression of genes 
```{r coexp2, cache=FALSE}
cr = cor(t(RESIDUAL.GENE_EXPRESSION))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Clustering residual data
```{r decompse.normalise.data2, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Dx, shape = Institution, size=RIN))
p <- p + theme_bw() + theme(legend.position="top") + scale_shape_manual(values = 1:9)
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=0)
p

# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c('Dx','Institution')])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(VOOM.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp[,c('Dx','Institution')]))
```

```{r temp1, include=F}
dev.off()
gc()
```

```{r get.gene.symbols, cache=TRUE, eval=T}
backgroundGenes = data.frame(ensembl_gene_id = rownames(COUNT))

# Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "jul2016.archive.ensembl.org", dataset = "hsapiens_gene_ensembl")

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", values = backgroundGenes$ensembl_gene_id,
                       mart = mart)

# Add gene names
RESIDUAL.GENE_EXPRESSION <- RESIDUAL.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  left_join(Ensemble2HGNC) %>%
  dplyr::select(ensembl_gene_id, hgnc_symbol, one_of(rownames(COVARIATES)))

VOOM.ADJUSTED.GENE_EXPRESSION$E <- VOOM.ADJUSTED.GENE_EXPRESSION$E %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  left_join(Ensemble2HGNC) %>%
  dplyr::select(ensembl_gene_id, hgnc_symbol, one_of(rownames(COVARIATES)))
```
### Differential expression analysis
Genes that are differentially expressed at an FDR <= 0.05 are
```{r diffExp, fig.height=8, fig.width=16}
# Fit contrast
contrast = makeContrasts(contrasts=c("DxControl-DxAFF", "DxControl-DxBP",
                                     "DxControl-DxSCZ", "DxAFF-DxBP",
                                     "DxAFF-DxSCZ", "DxBP-DxSCZ"),
                         levels = colnames(VOOM.ADJUSTED.FIT$coefficients))
FIT.CONTR = contrasts.fit(VOOM.ADJUSTED.FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)

# Get differnetial expression
DE = lapply(1:6, function(i, FIT, VOOM.ADJUSTED.GENE_EXPRESSION, Ensemble2HGNC){
  topTable(FIT.CONTR, coef=i, number = dim(VOOM.ADJUSTED.GENE_EXPRESSION$E)[1]) %>%
    rownameToFirstColumn('ensembl_gene_id') %>%
    left_join(Ensemble2HGNC) %>%
    dplyr::select(ensembl_gene_id, hgnc_symbol, logFC, AveExpr, t, P.Value, adj.P.Val, B)
}, FIT, VOOM.ADJUSTED.GENE_EXPRESSION, Ensemble2HGNC)
names(DE) = colnames(contrast)

DE = DE %>% rbindlist(idcol = 'Comparison')

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, Comparison) %>%
  group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id)))

tmp1 = DE %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::select(ensembl_gene_id, Comparison) %>%
  group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id)))

kable(full_join(tmp,tmp1))

p = ggplot(DE, aes(y = -log10(adj.P.Val), x = logFC)) + geom_point() + xlim(c(-1,1))
p = p + geom_hline(yintercept = -log10(0.05), color = 'red') 
p = p + geom_vline(xintercept = log2(1.2), color = 'red') 
p = p + geom_vline(xintercept = -log2(1.2), color = 'red') 
p = p + facet_grid(.~Comparison, scales = 'fixed')
p
```
### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Code
CODE <- Folder(name = "ACC - STAR Reprocessed - with TMM", parentId = parentId)
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'SampleID')
write.table(COVARIATES, file = 'CMC_ACC_Covariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = File('CMC_ACC_Covariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)

# Store logCPM
write.table(VOOM.ADJUSTED.GENE_EXPRESSION$E, file = 'CMC_ACC_logCPM.tsv', sep = '\t', row.names=F, quote=F)
COUNT_OBJ = File('CMC_ACC_logCPM.tsv', name = 'Raw counts (logCPM)', parentId = CODE$properties$id)
COUNT_OBJ = synStore(COUNT_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                     executed = thisFile, activityDescription = activityDescription)

# Store design matrix
write.table(DM1$design, file = 'CMC_ACC_Design.tsv', sep = '\t', row.names=F, quote=F)
DM_OBJ = File('CMC_ACC_Design.tsv', name = 'Design Matrix', parentId = CODE$properties$id)
DM_OBJ = synStore(DM_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                  executed = thisFile, activityDescription = activityDescription)

# Store residual gene expression
write.table(RESIDUAL.GENE_EXPRESSION, file = 'CMC_ACC_Expression.tsv', sep = '\t', row.names=F, quote=F)
EXP_OBJ = File('CMC_ACC_Expression.tsv', name = 'Normalised Adjusted Residual Expression (Dx added)', parentId = CODE$properties$id)
EXP_OBJ = synStore(EXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)

# Store differential expression results
write.table(DE, file = 'CMC_ACC_DiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('CMC_ACC_DiffExpression.tsv', name = 'Differential Expression Analysis', parentId = CODE$properties$id)
DEXP_OBJ = synStore(DEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
```
|  *Results*                |  *SynapseID*                     |
|  -------                  |   ---------                      |
|  Covariates               |  `r COV_OBJ$properties$id`       |
|  Raw counts (log CPM)     |  `r COUNT_OBJ$properties$id`     |
|  Design Matrix            |  `r DM_OBJ$properties$id`        |
|  Residual Expression      |  `r EXP_OBJ$properties$id`       |
|  Differential Expression  |  `r DEXP_OBJ$properties$id`      |