---
title: "Covariates analysis of UFL-Mayo-ISB RNASeq data of cerebellum tissue (realigned from synapse BAM with Picard-STAR-HTSeq)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(knit2synapse)
library(synapseClient)
library(knitr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./Mayo_CBE_RNASeq_Realigned_SynBAM_StarHTSeq.Rmd",
                                 parentId = "syn5570325",
                                 entityName = 'UFL_Mayo_ISB Cerebellum RNASeq RealignedFromSynBAM PicardStarHTSeq')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## It is assumed your working directory is where this file

### Clear R console screen output
cat("\014") 

## Load required libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(githubr)

library(ggplot2)
library(reshape2)
library(limma)
library(RColorBrewer)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(xlsx)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE}
# Synapse parameters
parentId = 'syn5570325';
activityName = 'Covariate analysis';
activityDescription = 'Covariate analysis of UFL_Mayo_ISB RNASeq data';

thisFileName <- 'Mayo_CBE_RNASeq_Realigned_SynBAM_StarHTSeq.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='AMPAD')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Get reprocessed paired end sample counts for cerebellum 
COUNT_ID <- 'syn7513807'
ALL_USED_IDs <- COUNT_ID
COUNT <- read.table(synGet(COUNT_ID)@filePath, header = T, row.names = 1, quote = '', check.names = F)

# Get clinical metadata
METADATA_ID <- 'syn5223705'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_ID
METADATA <- read.table(synGet(METADATA_ID)@filePath,sep=',',header=T, stringsAsFactors=F) %>%
  tidyr::separate(SampleID, c('Donor_ID', 'BrainRegion')) %>%
  dplyr::mutate(ID = paste(Donor_ID, BrainRegion, sep = '_')) %>%
  dplyr::rename(Gender = Sex, FLOWCELL = Flowcell)
```

### Data preprocessing
```{r preprocess.data,cache=FALSE, echo=TRUE}
# Fix metadata
METADATA$AgeAtDeath = gsub("_or_above", "", METADATA$AgeAtDeath)
METADATA$Diagnosis = gsub("Pathologic Aging","Pathologic_Aging", METADATA$Diagnosis)
METADATA$BrainRegion.Diagnosis = paste(METADATA$BrainRegion, METADATA$Diagnosis, sep = '.')
METADATA$RIN2 = METADATA$RIN^2
METADATA <- METADATA %>%
  filter(!is.na(Source)) %>%
  filter(!(ID %in% c('1923_CER', '11311_CER', '742_CER')))
```

```{r preprocess.data1, results='asis'}
# Match covariates to expression data
indToRetain = intersect(METADATA$ID, colnames(COUNT))
removedIDs = setdiff(colnames(COUNT), METADATA$ID)

COUNT = COUNT[,indToRetain]

rownames(METADATA) = METADATA$ID
METADATA = METADATA[indToRetain,]

METADATA %>% 
  group_by(BrainRegion, Diagnosis) %>% 
  summarise(count = n()) %>% 
  spread(BrainRegion, count) %>%
  kable()
```
Following sample are removed `r paste(removedIDs, collapse = ',')`

### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering}
FactorCovariates <- c("Source", "Gender", "FLOWCELL", "BrainRegion.Diagnosis")
ContCovariates <- c("RIN", "AgeAtDeath")

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
COVARIATES[,FactorCovariates] <- lapply(COVARIATES[,FactorCovariates],function(x){
  x <- sapply(x,function(y){str_replace_all(as.character(y),'[^[:alnum:]]','_')})
}) %>% data.frame()
rownames(COVARIATES) <- METADATA$ID

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)

# Add in RIN^2 values
COVARIATES$RIN2 = COVARIATES$RIN^2
ContCovariates = c(ContCovariates, 'RIN2')
```

Correlation/association between covariates at an FDR <= 0.1

```{r covariates.correlation, fig.width=8, fig.height=8}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.5, row.width=0.1)
```

Explore metatdata

```{r dataExplore, fig.width = 10, fig.height = 6}
my.theme <- theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1))

# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = RIN)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme

# Age of Death
p[[2]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = AgeAtDeath)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('Age At Death') + my.theme

multiplot(plotlist = p, cols = 2)
```
### CPM Normalisation
Preprocess counts matrix and metadata. Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of samples per BrainRegion.Diagnosis
```{r cpmnormalisation}
genesToAnalyze = dlply(METADATA, .(BrainRegion.Diagnosis), .fun = function(mtd, count){
  processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$ID %>% unique()],
                                                   MIN_GENE_CPM=1, 
                                                   MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
  processed.counts$filteredExprMatrix$genes
}, COUNT)
genesToAnalyze = unlist(genesToAnalyze) %>% unique()

PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, ], MIN_GENE_CPM=0, MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples

### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)
```{r initial.voom.normalisation, fig.height=5, fig.width=5}
# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)

# Set gene counts in specific samples that are deviating 3 sd from other samples to NA
log.mat = apply(VOOM.GENE_EXPRESSION$E, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  return((x < (mn-3*std.dev)) | (x > (mn+3*std.dev)))
}) %>% t
PROCESSED_COUNTS$filteredExprMatrix[log.mat] = NA

# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)
```
Coexpression of genes 
```{r coexp1, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(VOOM.GENE_EXPRESSION$E))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```
### Outlier analysis
Clustering of voom adjusted data (with NULL design)
```{r decompse.normalise.data, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'ID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Source, shape=BrainRegion.Diagnosis, size=AgeAtDeath))
p <- p + theme_bw() + theme(legend.position="top")
# p <- p + geom_text(aes(label= ID), size=4, hjust=0)
p
```
Tree based classification of samples
```{r decompse.normalise.data.1, fig.height=6, fig.width=10, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c('Source', 'Gender', 'FLOWCELL', 'BrainRegion.Diagnosis')])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(VOOM.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp[,c('Source', 'Gender', 'FLOWCELL', 'BrainRegion.Diagnosis')]))
```

```{r temp, include = F}
dev.off()
gc()
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, 
                                                 COVARIATES,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```

Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative design)
Since many covariates are correlated, re-normalising COVARIATES with an iterative design matrix
```{r iterative.norm}
# Primary variable of interest
postAdjustCovars = 'BrainRegion.Diagnosis'

# Assign residual covariates
residualSigCovars = preAdjustedSigCovars
significantCovars = residualSigCovars$significantCovars
covariatesEffects = residualSigCovars$Effects.significantCovars
covariatesEffects = covariatesEffects[setdiff(significantCovars, postAdjustCovars)]

postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 100){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))

  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]

  # Estimate correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F)

  # Fit a linear model
  FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION)

  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)

  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), postAdjustCovars)

  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                COVARIATES, 
                                                'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)

  # Add postadjusted covariates (if any)
  significantCovars = residualSigCovars$significantCovars
  covariatesEffects = residualSigCovars$Effects.significantCovars
  covariatesEffects = covariatesEffects[setdiff(significantCovars, postAdjustCovars)]

  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`

### Sanity check
```{r residualSigCovars.manual, fig.width=12, fig.height=8}
residualSigCovars[["PC_res"]][[2]]$plotData
```

### Residual calculation
Calculate weighted residuals and add back "Dx.Tissue" to the residuals
```{r varsToAddBack}
# Add variable of interest back to the residuals
varsToAddIn = grep("BrainRegion.Diagnosis", colnames(DM1$design), value = T)
RESIDUAL.GENE_EXPRESSION = RESIDUAL.GENE_EXPRESSION + 
  FIT$coefficients[,varsToAddIn] %*% t(DM1$design[,varsToAddIn])
```
Coexpression of genes 
```{r coexp2, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(RESIDUAL.GENE_EXPRESSION))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Clustering residual data
```{r decompse.normalise.data2, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION, scale.=T, center = T)

# Plot first 4 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'ID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2)) 
p <- p + geom_point(aes(color=Source, shape = BrainRegion.Diagnosis, size=RIN)) 
p <- p + theme_bw() + theme(legend.position="top")
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=1)
p
```

```{r decompse.normalise.data2.1, fig.height=8, fig.width=12, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c('Source', 'Gender', 'FLOWCELL', 'BrainRegion.Diagnosis')])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(VOOM.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp[,c('Source', 'Gender', 'FLOWCELL', 'BrainRegion.Diagnosis')]))
```

```{r temp1, include=F}
dev.off()
gc()
```

```{r get.gene.symbols, cache=TRUE}
# Define biomart object
mart <- useMart(host = "jul2016.archive.ensembl.org", dataset = "hsapiens_gene_ensembl", biomart = "ENSEMBL_MART_ENSEMBL")

RESIDUAL.GENE_EXPRESSION = rownameToFirstColumn(RESIDUAL.GENE_EXPRESSION, 'GeneID') %>%
  dplyr::mutate(gene.id = GeneID) %>%
  tidyr::separate(GeneID, c('ensembl_gene_id','position'),sep = '\\.')

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", values = RESIDUAL.GENE_EXPRESSION$ensembl_gene_id,
                       mart = mart)
RESIDUAL.GENE_EXPRESSION <- left_join(RESIDUAL.GENE_EXPRESSION, Ensemble2HGNC) %>%
  dplyr::select(gene.id, ensembl_gene_id, hgnc_symbol, one_of(rownames(COVARIATES)))
```

### Differential expression analysis
Genes that are differentially expressed at an FDR <= 0.05 are
```{r diffExp, fig.height=8, fig.width=20}
# Fit contrast
contrast = makeContrasts(contrasts=c("BrainRegion.DiagnosisCER_Control-BrainRegion.DiagnosisCER_AD",
                                     "BrainRegion.DiagnosisCER_Control-BrainRegion.DiagnosisCER_Pathologic_Aging",
                                     "BrainRegion.DiagnosisCER_Control-BrainRegion.DiagnosisCER_PSP",
                                     "BrainRegion.DiagnosisCER_Pathologic_Aging-BrainRegion.DiagnosisCER_AD",
                                     "BrainRegion.DiagnosisCER_PSP-BrainRegion.DiagnosisCER_AD",
                                     "BrainRegion.DiagnosisCER_PSP-BrainRegion.DiagnosisCER_Pathologic_Aging"),
                         levels = colnames(FIT$coefficients))
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)

# Get differnetial expression
DE = lapply(1:6, function(i, FIT, VOOM.ADJUSTED.GENE_EXPRESSION, Ensemble2HGNC){
  topTable(FIT.CONTR, coef=i, number = dim(VOOM.ADJUSTED.GENE_EXPRESSION$E)[1]) %>%
    rownameToFirstColumn('GeneID') %>%
    dplyr::mutate(gene.id = GeneID) %>%
    tidyr::separate(GeneID, c('ensembl_gene_id','position'),sep = '\\.') %>%
    left_join(Ensemble2HGNC) %>%
    dplyr::select(ensembl_gene_id, hgnc_symbol, logFC, AveExpr, t, P.Value, adj.P.Val, B)
}, FIT, VOOM.ADJUSTED.GENE_EXPRESSION, Ensemble2HGNC) 
names(DE) = colnames(contrast)

DE = DE %>% 
  rbindlist(idcol = 'Comparison', use.names = T, fill = T) %>%
  dplyr::mutate(Comparison = gsub('BrainRegion.Diagnosis', '', Comparison)) %>%
  tidyr::separate(Comparison, into = c('ref.state','to.state'), sep = '-') %>%
  tidyr::separate(ref.state, into = c('Tissue','Dx1'), sep = '_') %>%
  tidyr::separate(to.state, into = c('Tissue','Dx2'), sep = '_') %>%
  tidyr::unite(Comparison, Dx1, Dx2, sep = '_vs_') %>%
  dplyr::mutate(Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')))

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, Comparison, Tissue, Direction) %>%
  group_by(Comparison, Tissue, Direction) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05) 

tmp1 = DE %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::select(ensembl_gene_id, Comparison, Tissue, Direction) %>%
  group_by(Comparison, Tissue, Direction) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05_FC_1.2) 

kable(full_join(tmp %>%
                  dplyr::rename(DOWN_FDR_0_05 = DOWN, UP_FDR_0_05 = UP),
                tmp1 %>%
                  dplyr::rename(DOWN_FDR_0_05_FC_1.2 = DOWN, UP_FDR_0_05_FC_1.2 = UP)))

p = ggplot(DE, aes(y = -log10(adj.P.Val), x = logFC)) + geom_point() + xlim(c(-1,1))
p = p + geom_hline(yintercept = -log10(0.05), color = 'red') 
p = p + geom_vline(xintercept = log2(1.2), color = 'red') 
p = p + geom_vline(xintercept = -log2(1.2), color = 'red') 
p = p + facet_grid(Tissue+.~Comparison, scales = 'fixed')
p
```

### Comparison of reprocessed counts with source counts
```{r comp, fig.height=5, fig.width=5}
COUNT_OR = downloadFile('syn5201012')
rownames(COUNT_OR) = COUNT_OR$ensembl_id
COUNT_OR$ensembl_id = NULL

COUNT = COUNT %>%
  rownameToFirstColumn('GeneID') %>%
  dplyr::mutate(gene.id = GeneID) %>%
  tidyr::separate(GeneID, c('ensembl_gene_id','position'),sep = '\\.') %>%
  filter(ensembl_gene_id %in% rownames(COUNT_OR))
rownames(COUNT) = COUNT$ensembl_gene_id
COUNT = COUNT[intersect(rownames(COUNT), rownames(COUNT_OR)),intersect(colnames(COUNT), colnames(COUNT_OR))]
COUNT_OR = COUNT_OR[intersect(rownames(COUNT), rownames(COUNT_OR)),intersect(colnames(COUNT), colnames(COUNT_OR))]

cr = cor(COUNT, COUNT_OR, use = 'pairwise.complete.obs')
boxplot(diag(cr), main = 'Individual sample level correlation', ylab = 'Correlation', xlab = 'n=271')
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Code
CODE <- Folder(name = 'UFL_Mayo_ISB Cerebellum RNASeq RealignedFromSynBAM PicardStarHTSeq', parentId = parentId)
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'ID')
write.table(COVARIATES, file = 'MAYOcovariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = File('MAYOcovariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)
      
# Raw Expression
write.table(VOOM.ADJUSTED.GENE_EXPRESSION$E, file='MAYOlogCPM.tsv', sep = '\t', row.names=F, quote=F)
RAWCPM_OBJ = File('MAYOlogCPM.tsv', 
                  name = 'Raw logCPM',
                  parentId = CODE$properties$id)
RAWCPM_OBJ = synStore(RAWCPM_OBJ,
                      used = ALL_USED_IDs, 
                      activityName = activityName, 
                      executed = thisFile, 
                      activityDescription = activityDescription)

# Design Matrix
DM1$design = DM1$design[,!(colnames(DM1$design) %in% c('PMI'))]
write.table(DM1$design, file='MAYOdesign.tsv', sep = '\t', row.names=F, quote=F)
DM_OBJ = File('MAYOdesign.tsv', 
                  name = 'Design Matrix',
                  parentId = CODE$properties$id)
DM_OBJ = synStore(DM_OBJ,
                      used = ALL_USED_IDs, 
                      activityName = activityName, 
                      executed = thisFile, 
                      activityDescription = activityDescription)

# Store residual gene expression
write.table(RESIDUAL.GENE_EXPRESSION, file = 'MAYOexpression.tsv', sep = '\t', row.names=F, quote=F)
EXP_OBJ = File('MAYOexpression.tsv', name = 'Voom Adjusted Residual Expression (BrainRegion.Diagnosis added)', parentId = CODE$properties$id)
EXP_OBJ = synStore(EXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)

# Store differential expression results
write.table(DE, file = 'MAYOdiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('MAYOdiffExpression.tsv', name = 'Differential Expression Analysis', parentId = CODE$properties$id)
DEXP_OBJ = synStore(DEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)
```
|  *Results*                |  *SynapseID*                    |
|  -------                  |   ---------                     |
|  Covariates               |  `r COV_OBJ$properties$id`      |
|  Raw counts (logCPM)      |  `r RAWCPM_OBJ$properties$id`   |
|  Design Matrix            |  `r DM_OBJ$properties$id`       |
|  Residual Expression      |  `r EXP_OBJ$properties$id`      |
|  Differential Expression  |  `r DEXP_OBJ$properties$id`     |

### Source R Code
[This File](`r thisFile`)