---
title: "Covariate analysis of UFL-Mayo-ISB RNASeq"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(knit2synapse)
library(synapseClient)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./Mayo_Cerebellum_RNASeq.Rmd",
                                 parentId ="syn5570325",
                                 entityName = 'UFL_Mayo_ISB Cerebellum RNASeq')
```


```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file

## Load required libraries
library(synapseClient)
library(ggplot2)
library(reshape2)
library(data.table)
library(dplyr)
library(stringr)
library(limma)
library(knitr)
library(Biobase)
library(GEOquery)
library(RColorBrewer)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R", full.names=T)
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn5570325';
activityName = 'Covariate analysis';
activityDescription = 'Covariate analysis of UFL_Mayo_ISB cerebellum RNASeq data';

thisFileName <- 'Mayo_Cerebellum_RNASeq.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='AMPAD')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download expression data
COUNT_ID <- 'syn5201012';
ALL_USED_IDs <- COUNT_ID
COUNT_OBJ <- synGet(COUNT_ID)
COUNT <- read.table(COUNT_OBJ@filePath,header=T,sep='\t',check.names = F)

# Get clinical metadata
METADATA_ID <- 'syn5223705'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_ID
METADATA_OBJ <- synGet(METADATA_ID)
METADATA <- read.table(METADATA_OBJ@filePath,sep=',',header=T, stringsAsFactors=F)
```

### Data preprocessing
```{r preprocess.data,cache=TRUE, echo=TRUE}
# Fix metadata
METADATA$AgeAtDeath = gsub("_or_above", "", METADATA$AgeAtDeath)
METADATA <- METADATA %>%
  dplyr::filter(!(SampleID %in% c("742_CER", "11386_CER","11311_CER","1923_CER","1147_CER")))
```
Following five samples were considered outliers and removed 742_CER, 11386_CER, 11311_CER, 1923_CER, 1147_CER
```{r preprocess.data1,cache=TRUE}
indToRetain = intersect(METADATA$SampleID, colnames(COUNT))

EXPR <- COUNT;
EXPR <- EXPR[, colnames(EXPR) %in% indToRetain]

# Match covariates to expression data
rownames(METADATA) = METADATA$SampleID
METADATA = METADATA[rownames(METADATA) %in% indToRetain,]

EXPR = EXPR[,indToRetain]
COUNT = COUNT[rownames(EXPR),colnames(EXPR)]
METADATA = METADATA[indToRetain,]
```
Dorsolateral prefrontal cortex of `r dim(EXPR)[2]` subjects from the Mayo cohorts were obtained from AMP-AD portal

### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering, cache=TRUE}
FactorCovariates <- c("Source", "Diagnosis",  "Sex",  "ApoE", "Flowcell")
ContCovariates <- c("RIN", "AgeAtDeath")
  
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- METADATA$SampleID

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
Covariate correlation
```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.1)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```

### Gene Filtering
Retain genes with atleast 50% of the samples have >= 1 gene counts
```{r gene.filtering, results='asis', echo=FALSE}
# Remove genes
EXPR <- getGeneFilteredGeneExprMatrix(EXPR, 
                                      MIN_GENE_CPM=1,
                                      MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5,
                                      verbose=T)
```

### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)
```{r initial.voom.normalisation}
# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(EXPR$filteredExprMatrix, design=NULL, plot=T)

# Initial design matrix with covariates under consideration
DM = getDesignMatrix(COVARIATES, Intercept = F)
```
Clustering of initial normalised data (with NULL design)
```{r decompse.normalise.data, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, METADATA, by="SampleID")

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=factor(Diagnosis), shape=as.factor(Source), size=as.numeric(AgeAtDeath)))
p <- p + theme_bw() + theme(legend.position="top") 
p <- p + geom_text(aes(label= SampleID), size=4, hjust=0)
p

# Tree based clustering
sampleTrees = flashClust::hclust(dist(t(VOOM.GENE_EXPRESSION$E)), method = "average")
tree = WGCNA::cutreeStatic(sampleTrees, cutHeight = 200, minSize=3)

WGCNA::plotDendroAndColors(sampleTrees, 
                    labels2colors(cbind(tree,COVARIATES)),
                    groupLabels = c('Cluster',colnames(COVARIATES)))
collectGarbage()
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, results='asis', echo=FALSE}
# Find correlation between PC's of gene expression with covariates
DESIGN = getDesignMat2Fact(DM$design,FactorCovariates)
colnames(DESIGN) = substr(colnames(DESIGN),1,50)
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, DESIGN,
                                                 'NULL design(voom-normalized)',
                                                 isKeyPlot=TRUE)

# Find significant covariates
adjustCovars = designMatVarsToCovars(getCovariatesMap(DM),
                                     preAdjustedSigCovars$significantCovars)
```
Significant covariates to adjust at FDR 0.1 are `r paste(gsub('_','\\\\_',adjustCovars), collapse= ',')`
```{r preAdjustedSigCovars.NULL.ALL, fig.width=25, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```
### Normalisation (iterative)
Since many covariates are correlated, re-normalising COVARIATES with an iterative design matrix.
```{r iterative.adjusted.voom.normalisation, cache=TRUE, results='asis', echo=FALSE}
postAdjustCovars = c('Diagnosis')
 
# Assign residual covariates
residualSigCovars = preAdjustedSigCovars

significantCovars = designMatVarsToCovars(getCovariatesMap(DM), 
                                          residualSigCovars$significantCovars)

covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),
                                         residualSigCovars$Effects.significantCovars)

covariatesEffects = covariatesEffects[setdiff(significantCovars, postAdjustCovars)]

postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 100){
  writeLines(paste('Using following covariates in the model:', 
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    
  # Estimate voom weights
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(EXPR$filteredExprMatrix, design=DM1$design, plot=F)
  
  # Estimate voom weights
  FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, 
              design=DM1$design, 
              weights = VOOM.ADJUSTED.GENE_EXPRESSION$weights)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), postAdjustCovars)
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                DESIGN,
                                                'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  significantCovars = designMatVarsToCovars(getCovariatesMap(DM), residualSigCovars$significantCovars)
  covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),residualSigCovars$Effects.significantCovars)
  covariatesEffects = covariatesEffects[setdiff(significantCovars,postAdjustCovars)]
  
  postAdjustCovars = c(postAdjustCovars,names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
modelStr <-paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=','), 'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`

```{r adjusted.voom.normalisation, results='asis', echo=FALSE}
# Post adjusted design matrix
DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F], Intercept = F)
tmp = linColumnFinder(DM1$design)
writeLines('Degenerate variables are:')
writeLines(tmp$relations)
DM1$design = DM1$design[,tmp$indepCols] 

# Recalculate weights with new design
VOOM.ADJUSTED.GENE_EXPRESSION = voom(EXPR$filteredExprMatrix, 
                                     design=DM1$design, plot=T)

# Linear model with new weights and design
VOOM.ADJUSTED.FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, 
                          design=DM1$design, 
                          weights = VOOM.ADJUSTED.GENE_EXPRESSION$weights)  
```
### Sanity Check
Residualise significant covariates using a linear model and find correlation between PCA of residuals with covariates
```{r calculate.residuals, include=FALSE}
# Residuals after normalisation
RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,
                                              VOOM.ADJUSTED.GENE_EXPRESSION$E)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                              DESIGN,
                                              'residual matrix of all adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
```
```{r residualSigCovars.ALL, fig.width=25, fig.height=12}
residualSigCovars[["PC_res"]][[2]]$plotData
```
```{r addVarsBack}
varsToAddBackIn <- c("DiagnosisAD", "DiagnosisControl", "DiagnosisPSP", "DiagnosisPathologic Aging")
RESIDUAL.GENE_EXPRESSION = RESIDUAL.GENE_EXPRESSION + 
  VOOM.ADJUSTED.FIT$coefficients[,varsToAddBackIn] %*% t(data.matrix(DM1$design[,varsToAddBackIn]))
RESIDUAL.GENE_EXPRESSION = rownameToFirstColumn(RESIDUAL.GENE_EXPRESSION, 'ensembl_gene_id') 
```

```{r summary, results='asis', echo=FALSE}
writeLines('Number of samples in each category are')
tmp = as.data.frame(summary(COVARIATES$Diagnosis))
colnames(tmp) = 'No. of samples'
kable(tmp)
```

```{r get.gene.symbols, cache=TRUE}
# Define biomart object
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", values = RESIDUAL.GENE_EXPRESSION$ensembl_gene_id,
                       mart = mart)
RESIDUAL.GENE_EXPRESSION <- left_join(RESIDUAL.GENE_EXPRESSION, Ensemble2HGNC) %>%
  dplyr::select(ensembl_gene_id, hgnc_symbol, one_of(rownames(COVARIATES)))
```
### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Code
CODE <- Folder(name = 'UFL_Mayo_ISB Cerebellum RNASeq', parentId = parentId)
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'SampleID')
write.table(COVARIATES, file = 'covariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = File('covariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)
      
# Store covariates
write.table(RESIDUAL.GENE_EXPRESSION, file = 'expression.tsv', sep = '\t', row.names=F, quote=F)
EXP_OBJ = File('expression.tsv', name = 'Voom Adjusted Residual Expression (Diagnosis added)', parentId = CODE$properties$id)
EXP_OBJ = synStore(EXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)
```
|  *Results*            |  *SynapseID*                    |
|  -------              |   ---------                     |
|  Covariates           |  `r COV_OBJ$properties$id`      |
|  Residual Expression  |  `r EXP_OBJ$properties$id`      |