---
title: "Covariate analysis of UFL-Mayo-ISB RNASeq (both cerebellum and temporal cortex) samples from MayoBrainBank_Dickson alone"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(knit2synapse)
library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./Mayo_MayoBrainBankDickson_RNASeq.Rmd",
                                 parentId = "syn5570325",
                                 entityName = 'UFL_Mayo_ISB Mayo Brain Bank Dickson RNASeq')
```


```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## It is assumed your working directory is where this file

## Load required libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(reshape2)
library(stringr)
library(limma)
library(Biobase)
library(RColorBrewer)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(fpc)

library(doParallel)
library(foreach)

cl = makeCluster(14)
registerDoParallel(cl)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R", full.names=T)
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
dev.off()
```

```{r synapse.parameters, include=FALSE}
# Synapse parameters
parentId = 'syn5570325';
activityName = 'Covariate analysis';
activityDescription = 'Covariate analysis of UFL_Mayo_ISB RNASeq data';

thisFileName <- 'Mayo_MayoBrainBankDickson_RNASeq.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='AMPAD')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download expression data
COUNT_TC_ID <- 'syn4650257';
ALL_USED_IDs <- COUNT_TC_ID
COUNT_TC <- read.table(synGet(COUNT_TC_ID)@filePath, header=T, sep='\t', check.names = F)

# Download expression data
COUNT_CE_ID <- 'syn5201012';
ALL_USED_IDs[length(ALL_USED_IDs)+1] <- COUNT_CE_ID
COUNT_CE <- read.table(synGet(COUNT_CE_ID)@filePath, header=T, sep='\t', check.names = F)

COUNT = plyr::join_all(list(COUNT_TC, COUNT_CE)) %>%
  as.data.frame

# Get clinical metadata
METADATA_TC_ID <- 'syn3817650'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_TC_ID
METADATA_TC <- read.table(synGet(METADATA_TC_ID)@filePath,sep=',',header=T, stringsAsFactors=F) %>%
  tidyr::separate(ID, c('Donor_ID', 'BrainRegion'), sep = '_') %>%
  dplyr::mutate(ID = paste(Donor_ID, BrainRegion, sep = '_'))

# Get clinical metadata
METADATA_CE_ID <- 'syn5223705'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_CE_ID
METADATA_CE <- read.table(synGet(METADATA_CE_ID)@filePath,sep=',',header=T, stringsAsFactors=F) %>%
  tidyr::separate(SampleID, c('Donor_ID', 'BrainRegion')) %>%
  dplyr::mutate(ID = paste(Donor_ID, BrainRegion, sep = '_')) %>%
  dplyr::rename(Gender = Sex, FLOWCELL = Flowcell)

METADATA = rbindlist(list(METADATA_TC, METADATA_CE), use.names = T, fill = T) %>%
  as.data.frame
```

### Data preprocessing
```{r preprocess.data,cache=FALSE, echo=TRUE}
# Fix metadata
METADATA$AgeAtDeath = gsub("_or_above", "", METADATA$AgeAtDeath)
METADATA$Diagnosis = gsub("Pathologic Aging","Pathologic_Aging", METADATA$Diagnosis)
METADATA$BrainRegion.Diagnosis = paste(METADATA$BrainRegion, METADATA$Diagnosis, sep = '.')
METADATA$RIN2 = METADATA$RIN^2
METADATA <- METADATA %>%
  dplyr::filter(!(ID %in% c("742_CER", "11386_CER","11311_CER","1923_CER","11477_CER", "11344_CER", 
                            "1923_TCX", "11423_TCX", "1940_TCX", "7095_TCX", "11408_TCX"))) %>%
  filter(Source == 'MayoBrainBank_Dickson')
```
Following samples were considered outliers and removed based on PCA and/or eucledian distance:
```{r preprocess.data1}
# Match covariates to expression data
indToRetain = intersect(METADATA$ID, colnames(COUNT))

COUNT <- COUNT[, c('ensembl_id', intersect(colnames(COUNT), indToRetain))]

rownames(METADATA) = METADATA$ID
METADATA = METADATA[rownames(METADATA) %in% indToRetain,]

COUNT = COUNT[, c('ensembl_id', indToRetain)]
METADATA = METADATA[indToRetain,]

rownames(COUNT) = COUNT$ensembl_id
COUNT$ensembl_id = NULL
```
`r dim(COUNT)[2]` samples from `r length(unique(METADATA$Donor_ID))` subjects were obtained from the Mayo cohorts in AMP-AD portal

```{r diagnosisPlots, fig.height=10, fig.width = 15}
# RIN
p = list()
p[[1]] = ggplot(METADATA, aes(x = Diagnosis, y = RIN, fill = Tissue)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + theme(legend.position = 'top')

# AgeAtDeath
p[[2]] = ggplot(METADATA, aes(x = Diagnosis, y = as.numeric(AgeAtDeath), fill = Tissue)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AgeAtDeath') + theme(legend.position = 'top')

multiplot(plotlist = p, cols = 2)

# Source
vcd::mosaic(~ Source + Diagnosis + Tissue, data = METADATA)
```

### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering}
FactorCovariates <- c("Donor_ID", "Gender", "FLOWCELL", "BrainRegion.Diagnosis")
ContCovariates <- c("RIN", "RIN2", "AgeAtDeath")
  
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- METADATA$ID

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
Covariate correlation
```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = CovariateAnalysis::getAssociationStatistics(COVARIATES, PVAL = 0.05)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.4, row.width=0.1)
```

### Gene Filtering
Retain genes with atleast 50% of the samples have >= 1 gene counts
```{r gene.filtering, results='asis', echo=FALSE}
# Remove genes
EXPR <- getGeneFilteredGeneExprMatrix(COUNT, 
                                      MIN_GENE_CPM=1,
                                      MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5,
                                      verbose=T)
```

### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)
```{r initial.voom.normalisation}
# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(EXPR$filteredExprMatrix, design=NULL, plot=T)

# Initial design matrix with covariates under consideration
DM = getDesignMatrix(COVARIATES, Intercept = F)
DESIGN = getDesignMat2Fact(DM$design,FactorCovariates)
colnames(DESIGN) = substr(colnames(DESIGN),1,50)
```

Clustering of initial normalised data (with NULL design)
```{r decompse.normalise.data1, fig.height=15, fig.width=10, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E)

# Plot first 2 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2],
                       PC3=PC$rotation[,3],
                       PC4=PC$rotation[,4])

plotdata <- left_join(plotdata, METADATA, by="ID")

pl = list()
p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(shape=factor(Source), color=as.factor(Tissue), size=as.numeric(AgeAtDeath)))
p <- p + theme_bw() + theme(legend.position="top") +scale_shape_manual(values = 1:2)
p <- p + geom_text(aes(label= ID), size=4, hjust=0)
pl[[1]] <- p

p <- ggplot(plotdata, aes(x=PC3, y=PC4))
p <- p + geom_point(aes(shape=factor(Source), color=as.factor(Diagnosis), size=as.numeric(AgeAtDeath)))
p <- p + theme_bw() + theme(legend.position="top") +scale_shape_manual(values = 1:2)
# p <- p + geom_text(aes(label= ID), size=4, hjust=0)
pl[[2]] <- p

multiplot(plotlist = pl, cols = 1)
```
```{r decompse.normalise.data2, fig.height=9, fig.width=25, results='asis'}
# Tree based clustering
sampleTrees = flashClust::hclust(dist(t(VOOM.GENE_EXPRESSION$E)), method = "average")
tree = WGCNA::cutreeStatic(sampleTrees, cutHeight = 135, minSize=10)

COVARIATES.tmp = COVARIATES
COVARIATES.tmp[,FactorCovariates] = lapply(COVARIATES.tmp[,FactorCovariates], as.numeric)

WGCNA::plotDendroAndColors(sampleTrees, 
                           labels2colors(cbind(tree,COVARIATES.tmp)),
                           groupLabels = c('Cluster',colnames(COVARIATES.tmp)))
dev.off()

tmp = paste('Samples in cluster 0 are', paste(sampleTrees$labels[tree == 0], collapse = ','))
tmp = base::gsub('_','\\\\_',tmp)
writeLines(tmp)
tmp = paste('Samples in cluster 3 are', paste(sampleTrees$labels[tree == 3], collapse = ','))
tmp = base::gsub('_','\\\\_',tmp)
writeLines(tmp)
```
### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, results='asis', echo=FALSE}
# Find correlation between PC's of gene expression with covariates
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, COVARIATES,
                                                 'NULL design(voom-normalized)',
                                                 isKeyPlot=TRUE)

# Find significant covariates
adjustCovars = preAdjustedSigCovars$significantCovars
```
Significant covariates to adjust at FDR 0.1 are `r paste(gsub('_','\\\\_',adjustCovars), collapse= ',')`
```{r preAdjustedSigCovars.NULL.ALL, fig.width=25, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```
### Normalisation (iterative)
Since many covariates are correlated, re-normalising COVARIATES with an iterative design matrix.
```{r iterative.adjusted.voom.normalisation, results='asis', echo=FALSE}
postAdjustCovars = c('BrainRegion.Diagnosis')
randomCovariates = 'Donor_ID'

# Assign residual covariates
residualSigCovars = preAdjustedSigCovars
significantCovars = residualSigCovars$significantCovars
covariatesEffects = residualSigCovars$Effects.significantCovars
covariatesEffects = covariatesEffects[setdiff(significantCovars, c(postAdjustCovars, randomCovariates))]

postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))

loopCount = 0 
while(length(significantCovars)!=0 && loopCount <= 100){
  writeLines(paste('Using following covariates in the model:', 
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    
  # Estimate voom weights
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(EXPR$filteredExprMatrix, design=DM1$design, plot=F)
  
  # Calculate correlation values of random effects
  correlation = parallelDuplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION, block = COVARIATES$Donor_ID)
  
  # Re-estimate voom weights
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(EXPR$filteredExprMatrix, design=DM1$design, plot=F, 
                                       block = COVARIATES$Donor_ID, correlation = correlation$cor)
  
  # Fit a mixed effect linear model
  FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), c(postAdjustCovars, randomCovariates))
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                COVARIATES,
                                                'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  significantCovars = residualSigCovars$significantCovars
  covariatesEffects = residualSigCovars$Effects.significantCovars
  covariatesEffects = covariatesEffects[setdiff(significantCovars,c(postAdjustCovars, randomCovariates))]
  
  postAdjustCovars = c(postAdjustCovars,names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
modelStr <-paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=','), 'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr, 'and Donor_ID as random effects')
```
`r tmp`

### Sanity Check
Residualise significant covariates using a linear model and find correlation between PCA of residuals with covariates
```{r residualSigCovars.ALL, fig.width=25, fig.height=12}
residualSigCovars[["PC_res"]][[2]]$plotData
```
Adding "BrainRegion.Diagnosis" back to the residuals
```{r addVarsBack}
varsToAddBackIn <-  grep("BrainRegion.Diagnosis", colnames(FIT$coefficients), value = T)
RESIDUAL.GENE_EXPRESSION = RESIDUAL.GENE_EXPRESSION + 
  t(data.matrix(DM1$design[,varsToAddBackIn]) %*% t(FIT$coefficients[,varsToAddBackIn]))
RESIDUAL.GENE_EXPRESSION = rownameToFirstColumn(RESIDUAL.GENE_EXPRESSION, 'ensembl_gene_id') 
```

### Clustering residual gene expression
```{r decompse.normalise.data3, fig.height=10, fig.width=9, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION[,-(1)], scale.=T)

# Tree based clustering
sampleTrees = flashClust::hclust(dist(t(RESIDUAL.GENE_EXPRESSION[,-(1)])), method = "average")

# Cluster expression
clust = cutreeStatic(sampleTrees, cutHeight = 100, minSize = 10)

# Plot first 2 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2],
                       PC3=PC$rotation[,3], 
                       PC4=PC$rotation[,4],
                       clust = clust)

plotdata <- left_join(plotdata, METADATA, by="ID")

pl = list()
p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(shape=factor(Source), color=as.factor(Tissue), size=as.numeric(AgeAtDeath)))
p <- p + theme_bw() + theme(legend.position="top") +scale_shape_manual(values = 1:2)
p <- p + geom_text(aes(label= ID), size=4, hjust=0)
pl[[1]] <- p

p <- ggplot(plotdata, aes(x=PC3, y=PC4))
p <- p + geom_point(aes(shape=factor(Source), color=as.factor(Diagnosis), size=as.numeric(AgeAtDeath)))
p <- p + theme_bw() + theme(legend.position="top") +scale_shape_manual(values = 1:2)
# p <- p + geom_text(aes(label= ID), size=4, hjust=0)
pl[[2]] <- p

multiplot(plotlist = pl, cols = 1)

names(clust) = sampleTrees$labels
clust = clust[rownames(COVARIATES)]
```
```{r decompse.normalise.data4, fig.height=10, fig.width=25, results='asis'}
# Tree based clustering
COVARIATES.tmp = COVARIATES
COVARIATES.tmp[, FactorCovariates] = sapply(COVARIATES.tmp[,FactorCovariates], as.numeric)
plotDendroAndColors(sampleTrees, 
                    colors = labels2colors(cbind(data.frame(clust = clust), COVARIATES.tmp)), 
                    abHeight = 0.80, 
                    main = "Sample dendrogram",
                    groupLabels = colnames(COVARIATES))

tmp = paste('Samples in cluster 0 are', paste(sampleTrees$labels[tree == 0], collapse = ','))
tmp = base::gsub('_','\\\\_',tmp)
writeLines(tmp)
tmp = paste('Samples in cluster 3 are', paste(sampleTrees$labels[tree == 3], collapse = ','))
tmp = base::gsub('_','\\\\_',tmp)
writeLines(tmp)
tmp = paste('Samples in cluster 4 are', paste(sampleTrees$labels[tree == 4], collapse = ','))
tmp = base::gsub('_','\\\\_',tmp)
writeLines(tmp)
```

```{r summary, results='asis', echo=FALSE}
writeLines('Number of samples in each category are')
tmp = data.frame(summary(COVARIATES$BrainRegion.Diagnosis))
colnames(tmp) = "No. of samples"
kable(tmp)
```

```{r get.gene.symbols}
# Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = "www.ensembl.org")

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", values = RESIDUAL.GENE_EXPRESSION$ensembl_gene_id,
                       mart = mart)
RESIDUAL.GENE_EXPRESSION <- left_join(RESIDUAL.GENE_EXPRESSION, Ensemble2HGNC) %>%
  dplyr::select(ensembl_gene_id, hgnc_symbol, one_of(rownames(COVARIATES)))
```
### Differential expression analysis
Genes that are differentially expressed at an FDR <= 0.05 are
```{r diffExp, fig.height=25, fig.width=25}
# Fit contrast
contrast = makeContrasts(contrasts=c(
  "BrainRegion.DiagnosisCER.Control-BrainRegion.DiagnosisCER.AD",
  "BrainRegion.DiagnosisCER.Control-BrainRegion.DiagnosisCER.PSP",
  "BrainRegion.DiagnosisCER.PSP-BrainRegion.DiagnosisCER.AD",
  "BrainRegion.DiagnosisTCX.Control-BrainRegion.DiagnosisTCX.AD",
  "BrainRegion.DiagnosisTCX.Control-BrainRegion.DiagnosisTCX.PSP",
  "BrainRegion.DiagnosisTCX.PSP-BrainRegion.DiagnosisTCX.AD"), 
                         levels = colnames(FIT$coefficients))
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)

# Get differnetial expression
DE = list()
p = list()
for (i in 1:6){
  DE[[i]] = topTable(FIT.CONTR, coef=i, number = dim(VOOM.ADJUSTED.GENE_EXPRESSION$E)[1])
  DE[[i]] = rownameToFirstColumn(DE[[i]], 'ensembl_gene_id') %>%
    left_join(Ensemble2HGNC)
  p[[i]] <- ggplot(DE[[i]], aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()
  p[[i]] <- p[[i]] + geom_hline(yintercept = -log10(0.05), color = 'red') 
  p[[i]] <- p[[i]] + ggtitle(gsub('BrainRegion.Diagnosis','',colnames(contrast)[i]))
}
multiplot(plotlist=p, cols = 3)
names(DE) = gsub('BrainRegion.Diagnosis','',colnames(contrast))

DE = plyr::ldply(DE, .id = 'Comparison')
tmp = DE %>% 
  dplyr::filter(adj.P.Val <= 0.05) %>%
  group_by(Comparison) %>%
  dplyr::summarize(nDiffExpGenes = length(unique(hgnc_symbol)))
kable(tmp)
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Code
CODE <- Folder(name = 'UFL_Mayo_ISB Mayo Brain Bank Dickson RNASeq', parentId = parentId)
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'ID')
write.table(COVARIATES, file = 'covariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = File('covariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)
      
# Store covariates
write.table(RESIDUAL.GENE_EXPRESSION, file = 'expression.tsv', sep = '\t', row.names=F, quote=F)
EXP_OBJ = File('expression.tsv', name = 'Voom Adjusted Residual Expression (BrainRegion.Diagnosis added)', parentId = CODE$properties$id)
EXP_OBJ = synStore(EXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)

# Store differential expression results
write.table(DE, file = 'diffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('diffExpression.tsv', name = 'Differential Expression Analysis', parentId = CODE$properties$id)
DEXP_OBJ = synStore(DEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)
stopCluster(cl)
```
|  *Results*            |  *SynapseID*                    |
|  -------              |   ---------                     |
|  Covariates           |  `r COV_OBJ$properties$id`      |
|  Residual Expression  |  `r EXP_OBJ$properties$id`      |
|  Differential Expression  |  `r DEXP_OBJ$properties$id`      |

### Source R Code
[This File](`r thisFile`)