---
title: "Meta analysis of differential expression results from AMP-AD (Control vs AD)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knit2synapse)
synapseLogin()

knit2synapse::knitToFolderEntity(file = "./MetaAnalysis.Rmd", 
                                 parentId ="syn5569102",
                                 entityName = 'MetaAnalysis of Control vs AD from all three brain banks')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(RColorBrewer)
library(ggplot2)
library(ComplexHeatmap)

library(RankProd)
library(githubr)

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r fxns}
# Utility function for combining p-values using Fishers inverse chi-square method
fishersPvalue <- function(pvalue){
  pvalue = -2*log(pvalue, base = exp(1))
  S <- 0
  N <- dim(pvalue)[2]
  
  correlation <- WGCNA::bicor(pvalue,use = 'pairwise.complete.obs');
  
  for (i in 1:(N-1))
    for (j in (i+1):N){
      if (correlation[i,j] > 0){
        S <- S + correlation[i,j]*(3.25+0.75*correlation[i,j])
      } else if (correlation[i,j] <= 0){
        S <- S + correlation[i,j]*(3.27+0.71*correlation[i,j])
      }
    }
  
  combinedPvalue.chi <- rowSums(pvalue)*(4*N)/(4*N+2*S)
  combinedPvalue <- sapply(combinedPvalue.chi, function(x,N){
    pchisq(x, df = 2*N, lower.tail = FALSE)
  },N)
  names(combinedPvalue) = rownames(pvalue)
  
  return(combinedPvalue)
}
```

### Get covariates/sample metadata from synapse
```{r covariates}
covariates.id = c('ROSMAP' = 'syn6114444',
                  'MSBB' = 'syn6131421',
                  'MAYO' = 'syn6128267')

covariates = lapply(covariates.id, downloadFile)

covariates$ROSMAP = covariates$ROSMAP %>%
  dplyr::select(Sampleid, msex, cogdx) %>%
  dplyr::rename(ID = Sampleid, Gender = msex, Status = cogdx) %>%
  dplyr::mutate(Gender = factor(Gender, labels = c('1' = 'MALE','2'='FEMALE')),
                Status = factor(Status, labels = c('1' = 'NCI', '2' = 'MCI', '3' = 'MCI.CI', 
                                                   '4' = 'AD', '5' = 'AD.CI', '6' = 'OD')),
                BrainRegion = 'DLPFC') %>%
  dplyr::filter(Status %in% c('NCI', 'AD'))

covariates$MSBB = covariates$MSBB %>%
  dplyr::select(SampleId, SEX, BrainRegion.Diagnosis) %>%
  tidyr::separate(BrainRegion.Diagnosis, c('BrainRegion', 'Dx'), sep = '\\.') %>%
  dplyr::rename(ID = SampleId, Gender = SEX, Status = Dx) %>%
  dplyr::mutate(Gender = factor(Gender, labels = c('M' = 'MALE','F'='FEMALE')),
                Status = factor(Status),
                BrainRegion = factor(BrainRegion, labels = c('BM_10' = 'FP', 'BM_22' = 'STG', 
                                                             'BM_36' = 'PHG', 'BM_44' = 'IFG'))) %>%
  dplyr::filter(Status %in% c('ND', 'SD'))

covariates$MAYO = covariates$MAYO %>%
  dplyr::select(ID, Gender, BrainRegion.Diagnosis) %>%
  tidyr::separate(BrainRegion.Diagnosis, c('BrainRegion', 'Dx'), sep = '\\.') %>%
  dplyr::rename(Status = Dx) %>%
  dplyr::mutate(Gender = factor(Gender, labels = c('M' = 'MALE','F'='FEMALE')),
                Status = factor(Status)) %>%
  dplyr::filter(Status %in% c('Control', 'AD'))

covariates = rbindlist(covariates, use.names = T, fill = T, idcol = 'Study')


```

### Get logCPM counts from synapse
```{r counts}
logcpm.id = c('ROSMAP' = 'syn6114447' ,
              'MSBB' = 'syn6131419',
              'MAYO' = 'syn6128269')

logcpm = lapply(logcpm.id, downloadFile) %>%
  lapply(function(x){
    x = dplyr::select(x, -hgnc_symbol)
    }) %>%
  join_all(type = 'full') %>%
  dplyr::select(Gene.ID, ensembl_gene_id, one_of(covariates$ID)) 
logcpm[is.na(logcpm)] = dplyr::select(logcpm, -Gene.ID, -ensembl_gene_id) %>%
  min
```

### Download differential expression data
```{r download.data}
diffexp.id = c('ROSMAP' = 'syn6114453',
               'MSBB' = 'syn6131423',
               'MAYO' = 'syn6128273')

diffexp = lapply(diffexp.id, downloadFile)

diffexp$ROSMAP = diffexp$ROSMAP %>%
  filter(Comparison == 'cogdx1-cogdx4') %>%
  dplyr::select(Gene.ID, logFC, P.Value, adj.P.Val) %>%
  dplyr::mutate(Tissue = 'DLPFC', Contrast = 'NCI-AD')
   
diffexp$MSBB = diffexp$MSBB %>%
  dplyr::filter(Name == 'ND-SD') %>%
  dplyr::select(Gene.ID, logFC, P.Value, adj.P.Val, BrainRegion, Name) %>%
  dplyr::mutate(BrainRegion = factor(BrainRegion, labels = c('BM_10' = 'FP', 'BM_22' = 'STG', 
                                                             'BM_36' = 'PHG', 'BM_44' = 'IFG'))) %>%
  dplyr::rename(Tissue = BrainRegion, Contrast = Name)
  
diffexp$MAYO = diffexp$MAYO %>%
  dplyr::filter(Name == 'Control-AD') %>%
  dplyr::select(genes, logFC, P.Value, adj.P.Val, BrainRegion, Name) %>%
  dplyr::rename(Tissue = BrainRegion, Gene.ID = genes, Contrast = Name)

diffexp = rbindlist(diffexp, use.names = T, fill = T, idcol = 'Study') %>%
  tidyr::separate(Gene.ID, c('ensembl_gene_id','position'),sep = '\\.')

diffexp$DE = 'NONE'
diffexp$DE[diffexp$adj.P.Val <= 0.05 & diffexp$logFC >= log2(1.2)] = 'UP'
diffexp$DE[diffexp$adj.P.Val <= 0.05 & diffexp$logFC <= -log2(1.2)] = 'DOWN'

pl = list()
for (study in c('ROSMAP', 'MSBB', 'MAYO')){
  p = filter(diffexp, Study == study) %>%
    ggplot(aes(x = logFC, y = -log10(adj.P.Val), color = DE)) 
  p = p + scale_colour_manual(values = c("green","grey", "red"))
  p = p + geom_point() + theme(legend.position = 'None')
  p = p + facet_grid(Study~.+Tissue)
  pl[[study]] = p
}
multiplot(plotlist = pl, cols = 1)
```

### Get hgnc_symbol mapping from biomaRt
```{r biomart}
# Define biomart object
mart <- useMart('ENSEMBL_MART_ENSEMBL', dataset = "hsapiens_gene_ensembl", host = 'dec2015.archive.ensembl.org')

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", 
                       values = unique(diffexp$ensembl_gene_id),
                       mart = mart) %>%
  data.table()

# Add hgnc_symbols to the logcpm matrix
logcpm = logcpm %>%
  left_join(Ensemble2HGNC)

# Add hgnc_symbols to the diffexp matrix
diffexp = diffexp %>%
  left_join(Ensemble2HGNC)
```

### Pathway enrichment
```{r pathway.enrichment}
# Differentially expressed in ROSMAP
## Up regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC >0, Tissue == 'DLPFC') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')
### Enrichr results are here: http://amp.pharm.mssm.edu/Enrichr/enrich?dataset=6x3h

## Down regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC<0, Tissue == 'DLPFC') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')
### Enrichr results are here: http://amp.pharm.mssm.edu/Enrichr/enrich?dataset=6x4e

# Differentially expressed in MSSM FP
## Up regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC >0, Tissue == 'FP') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')
### Enrichr results are here: http://amp.pharm.mssm.edu/Enrichr/enrich?dataset=6xcm

## Down regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC < 0, Tissue == 'FP') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')

# Differentially expressed in MSSM STG
## Up regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC > 0, Tissue == 'STG') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')

## Down regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC < 0, Tissue == 'STG') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')

# Differentially expressed in MSSM PHG
## Up regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC > 0, Tissue == 'PHG') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')
### Enrichr results are here: http://amp.pharm.mssm.edu/Enrichr/enrich?dataset=6xco

## Down regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC < 0, Tissue == 'PHG') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')
### Enrichr results are here: http://amp.pharm.mssm.edu/Enrichr/enrich?dataset=6xge

# Differentially expressed in MAYO CER
## Up regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC > 0, Tissue == 'CER') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')
### Enrichr results are here: 


## Down regulated
tmp = filter(diffexp, adj.P.Val <= 0.05, logFC < 0, Tissue == 'PHG') %>%
  dplyr::left_join(Ensemble2HGNC %>% as.data.table()) %>%
  filter(hgnc_symbol != '')
### Enrichr results are here: http://amp.pharm.mssm.edu/Enrichr/enrich?dataset=6xge
```


### Meta Analysis
```{r metaanalysis}
P.Value = diffexp %>% 
  # filter(Tissue %in% c('DLPFC', 'PHG', 'TCX')) %>%
  dplyr::select(ensembl_gene_id, P.Value, Tissue, Contrast) %>%
  tidyr::unite(Tissue.Contrast, Tissue, Contrast, sep = '_') %>%
  tidyr::spread(Tissue.Contrast, P.Value)
P.Value[is.na(P.Value)] = 1

min.pval = dplyr::select(P.Value, -ensembl_gene_id) %>%
  data.matrix %>% rowMin %>%
  p.adjust(method = 'BH') %>%
  data.frame(ensembl_gene_id = P.Value$ensembl_gene_id,
             min.pval = .)

max.pval = dplyr::select(P.Value, -ensembl_gene_id) %>%
  data.matrix %>% rowMax %>%
  p.adjust(method = 'BH') %>%
  data.frame(ensembl_gene_id = P.Value$ensembl_gene_id,
             max.pval = .)

fb.pval = dplyr::select(P.Value, -ensembl_gene_id) %>%
  data.matrix %>% fishersPvalue %>%
  p.adjust(method = 'BH') %>%
  data.frame(ensembl_gene_id = P.Value$ensembl_gene_id,
             fb.pval = .)

metaAnal = join_all(list(min.pval, max.pval, fb.pval), type = 'full')
tmp = metaAnal %>% gather(method, pval, -ensembl_gene_id) %>% group_by(method) %>% filter(pval <= 0.05, method == 'fb.pval') %>%
  left_join(Ensemble2HGNC) %>%
  dplyr::select(hgnc_symbol) %>%
  unique
```


```{r plotDE}
tmp = lapply(DEXP, function(x){
  x$observed.P.val = p.adjust(x$P.Value, 'fdr')
  x = arrange(x, observed.P.val)
  x = mutate(x, expected.P.val = seq(0,1,1/(dim(x)[1]-1)))
  x$dexp = 0;
  x$dexp[x$observed.P.val <= 0.05 & x$logFC > 0] = 1;
  x$dexp[x$observed.P.val <= 0.05 & x$logFC < 0] = -1;
  dplyr::select(x, observed.P.val, expected.P.val, dexp)
}) 

p = list()
for (i in 1:length(tmp)){
  p[[i]] = ggplot(tmp[[i]], aes(x = -log10(expected.P.val), y = -log10(observed.P.val))) + geom_point(aes(color = factor(dexp))) 
  p[[i]] = p[[i]] + scale_colour_manual(values=c('green','grey','red')) + theme(legend.position="none")
  x = -log10(tmp[[i]]$expected.P.val)
  x[4]-x[1]/
  p[[i]] = p[[i]] + geom_abline(slope=1, intercept=0) + ggtitle(names(tmp)[i])
}
multiplot(plotlist = p, cols = 4)
```

```{r uniqGenes}
GNames = DEXP[[1]]$hgnc_symbol
for (i in 2: length(DEXP)){
  GNames = intersect(GNames, DEXP[[i]]$hgnc_symbol)
}
GNames = GNames[!is.na(GNames)]
GNames = GNames[GNames != '']

DEXP = lapply(DEXP, function(dexp, GNames){
  dexp = dexp %>%
    dplyr::filter(hgnc_symbol %in% GNames) %>%
    group_by(hgnc_symbol) %>%
    dplyr::summarise(P.Value = min(P.Value)) %>%
    data.frame
  rownames(dexp) = dexp$hgnc_symbol
  dexp = dexp[GNames,]
  }, GNames)

DEXP = mapply(function(dexp, dataName){
  colnames(dexp) = c('hgnc_symbol', dataName)
  return(dexp)
}, DEXP, names(DEXP), SIMPLIFY = F)

DEXP = join_all(DEXP, by = "hgnc_symbol")
```
Overall `r dim(DEXP)[1]` genes in `r dim(DEXP)[2]-1` studies are analysed
### Differential expression analysis
```{r diffExp, results = 'asis'}
Adj.P.Value = apply(DEXP[,-(1)], 2, p.adjust, method = "fdr")
rownames(Adj.P.Value) = DEXP$hgnc_symbol

writeLines('Total number of genes that are differentially expressed at a FDR <= 0.05 are')
tmp = (data.frame(colSums(Adj.P.Value <= 0.05)))
colnames(tmp) = 'No. of Genes'
kable(tmp)

writeLines('Genes that are differentially expressed at a FDR <= 0.05 in 8/8 datasets')
paste(rownames(Adj.P.Value)[which(rowSums(Adj.P.Value<=0.05)==8)], collapse = ',')

writeLines('Genes that are differentially expressed at a FDR <= 0.05 in 7/8 datasets')
paste(rownames(Adj.P.Value)[which(rowSums(Adj.P.Value<=0.05)==7)], collapse = ',')

writeLines('Genes that are differentially expressed at a FDR <= 0.05 in 6/6 RNASeq datasets')
paste(rownames(Adj.P.Value)[which(rowSums(Adj.P.Value[,-(4:5)]<=0.05)==6)], collapse = ',')
```

### Minimum Pvalue meta analysis
```{r minP}
Pval = DEXP

Pmin = data.frame(hgnc_symbols = Pval[,1], Pmin = rowMin(data.matrix(Pval[,-(1)])))
Pmin$Pmin = p.adjust(Pmin$Pmin, method = 'fdr')

tmp = dplyr::filter(Pmin, Pmin <= 0.05) 
tmp = data.frame(No.of.Genes = dim(tmp)[1], 
                 GeneList = tmp %>%
                   dplyr::select(hgnc_symbols) %>%
                   unlist %>% paste(collapse = ','))
writeLines("Genes that are atleast differentially expressed at an FDR <= 0.05 in one study are")
kable(tmp)
```


```{r fisher.brown}
Pfb = data.frame(hgnc_symbols = Pval[,1], 
                 Pfb = p.adjust(fishersPvalue(Pval[,-(1)]), method = 'fdr'))

Pfb$Pfb = p.adjust(Pfb$Pfb, method = 'fdr')

tmp = dplyr::filter(Pfb, Pfb <= 0.05) 
tmp = data.frame(No.of.Genes = dim(tmp)[1], 
                 GeneList = tmp %>%
                   dplyr::select(hgnc_symbols) %>%
                   unlist %>% paste(collapse = ','))

writeLines("Genes that are differentially expressed in the meta analysis at an FDR <= 0.05 are")
kable(tmp)
```

### Store results in synapse
```{r synapse.store, include=FALSE, cache=FALSE}
thisFileName <- 'MetaAnalysis.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='metaAnal')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))

# Store consolidated data in synapse
CODE = Folder(name = 'MetaAnalysis of Control vs AD Data', parentId = "syn5569102")
CODE = synStore(CODE)

# Store covariates in synapse
write.table(covariates, file = 'ALLCovariates.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLCovariates.tsv', 
           name = 'Covariates', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')

# Store logcpm in synapse
write.table(logcpm, file = 'ALLLogCPM.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLLogCPM.tsv', 
           name = 'Log CPM (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')

# Store diffexp in synapse
write.table(diffexp, file = 'ALLDiffExp.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLDiffExp.tsv', 
           name = 'Differential Expression (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')
```