---
title: "Meta analysis of differential expression results from AMP-AD (Control vs AD, reprocessed RNASeq)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knit2synapse)
synapseLogin()

knit2synapse::knitToFolderEntity(file = "./MetaAnalysis.Rmd", 
                                 parentId ="syn5569102",
                                 entityName = 'MetaAnalysis of Control vs AD from all three brain banks (Reprocessed RNASeq)')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library(CovariateAnalysis) # install from th1vairam repo using devtools::install_github('th1vairam/CovariateAnalysis@devtools')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(RColorBrewer)
library(ggplot2)
library(ComplexHeatmap)
library(biomaRt)
library(Matrix)
library(org.Hs.eg.db)

library(limma)
library(edgeR)
library(RankProd)
library(githubr)

library(Vennerable)
library(colorfulVennPlot)

synapseLogin()

library(doParallel)
library(foreach)
library(doParallel)

clust = makeCluster(detectCores() - 2)
registerDoParallel(clust)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r fxns}
# Utility function for combining p-values using Fishers inverse chi-square method
fishersPvalue <- function(pvalue){
  pvalue = -2*log(pvalue, base = exp(1))
  S <- 0
  N <- dim(pvalue)[2]
  
  correlation <- WGCNA::bicor(pvalue,use = 'pairwise.complete.obs');
  
  for (i in 1:(N-1))
    for (j in (i+1):N){
      if (correlation[i,j] > 0){
        S <- S + correlation[i,j]*(3.25+0.75*correlation[i,j])
      } else if (correlation[i,j] <= 0){
        S <- S + correlation[i,j]*(3.27+0.71*correlation[i,j])
      }
    }
  
  combinedPvalue.chi <- rowSums(pvalue)*(4*N)/(4*N+2*S)
  combinedPvalue <- sapply(combinedPvalue.chi, function(x,N){
    pchisq(x, df = 2*N, lower.tail = FALSE)
  },N)
  names(combinedPvalue) = rownames(pvalue)
  
  return(combinedPvalue)
}
```

### Get covariates/sample metadata from synapse
```{r covariates}
covariates.ids = c(ROSMAP = 'syn8018344',
                   MSBB = 'syn8073656',
                   MAYO = 'syn8028561')
covariates = lapply(covariates.ids, function(id){
  x = downloadFile(id)
  colnames(x)[1] = 'SampleID'
  return(x)
})

covariates$ROSMAP = dplyr::filter(covariates$ROSMAP, cogdx %in% c(1,4)) %>%
  dplyr::mutate(Dx = factor(cogdx, levels = c(1,4), labels = c('1' = 'Control', '4' = 'AD')),
                Study = 'ROSMAP',
                BrainRegion = 'DLPFC') %>%
  dplyr::select(SampleID, Study, BrainRegion, Dx)

covariates$MSBB = dplyr::filter(covariates$MSBB, BrodmannArea.Dx %in% c('BM10.SD', 'BM22.SD', 'BM36.SD', 'BM44.SD', 'BM36.ND', 'BM10.ND', 'BM44.ND', 'BM22.ND')) %>%
  dplyr::select(SampleID, BrodmannArea.Dx) %>%
  tidyr::separate(BrodmannArea.Dx, c('BrainRegion', 'Dx')) %>%
  dplyr::mutate(Study = 'MSSM',
                Dx = factor(Dx, levels = c('ND','SD'), labels = c('ND' = 'Control', 'SD' = 'AD')),
                BrainRegion = factor(BrainRegion, labels = c('FP', 'STG', 'PHG', 'IFG'))) %>%
  dplyr::select(SampleID, Study, BrainRegion, Dx)

covariates$MAYO = dplyr::filter(covariates$MAYO, BrainRegion.Diagnosis %in% c('TCX.AD', 'TCX.Control', 'CER.AD', 'CER.Control')) %>%
  dplyr::select(SampleID, BrainRegion.Diagnosis) %>%
  tidyr::separate(BrainRegion.Diagnosis, c('BrainRegion', 'Dx')) %>%
  dplyr::mutate(Study = 'MAYO',
                Dx = factor(Dx, levels = c('Control','AD')),
                BrainRegion = factor(BrainRegion, levels = c('CER', 'TCX'))) %>%
  dplyr::select(SampleID, Study, BrainRegion, Dx)

covariates = covariates %>%
  rbindlist(use.names = T, fill = T)
```

### Get filtered gene.id and sample.id from synapse
```{r counts}
logcpm.ids = c('ROSMAP' = 'syn8018352',
               'MSBB' = 'syn8073666',
               'MAYO' = 'syn8028565')

logcpm = lapply(logcpm.ids, downloadFile) %>%
  lapply(function(x){
    x = dplyr::select(x, -hgnc_symbol, -ensembl_gene_id)
  }) %>%
  join_all(type = 'full') %>%
  dplyr::select(gene.id, one_of(covariates$SampleID)) 
logcpm[is.na(logcpm)] = min(logcpm[,-(1)], na.rm = T)
```

### Get raw counts from synapse
```{r raw.counts, eval = F}
raw.counts.ids = c('MAYO_CBE' = 'syn7513807',
                   'MSBB' = 'syn7994853',
                   'MAYO_TCX' = 'syn7987765',
                   'ROSMAP' = 'syn7750270')

lcpm = lapply(raw.counts.ids, function(id, lcpm){
  expr = read.table(synGet(id)@filePath, sep = '\t', check.names = F) 
  lib.size = calcNormFactors(expr, method = 'TMM')
  expr = voom(expr, lib.size = lib.size, design = NULL, plot = F, normalize.method = 'quantile')$E
  expr = expr[row.names(expr) %in% lcpm$gene.id, ] %>%
    rownameToFirstColumn('gene.id')
}, logcpm) %>%
  join_all(type = 'full')
```

### Get residual expression from synapse
```{r resid.expr}
# Get SVA adjusted data
resid.expr.id = c('ROSMAP' = 'syn6129372',
                  'MSBB' = 'syn7059289',
                  'MAYO' = 'syn6128634')

resid.expr = lapply(resid.expr.id, downloadFile) %>%
  lapply(function(x){
    x = dplyr::select(x, -hgnc_symbol)
    }) %>%
  join_all(type = 'full') %>%
  dplyr::select(Gene.ID, ensembl_gene_id, one_of(covariates$ID)) 
resid.expr[is.na(resid.expr)] = dplyr::select(resid.expr, -Gene.ID, -ensembl_gene_id) %>%
  min(na.rm=T)
```

### Get hgnc_symbol mapping from biomaRt
```{r biomart}
# Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = "dec2016.archive.ensembl.org")

# Get background gene ids
backgroundGenes = logcpm[,'gene.id', drop = F] %>% 
  dplyr::mutate(ID = gene.id) %>%
  tidyr::separate(ID, c('ensembl_gene_id','position'), sep = '\\.')

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", values = backgroundGenes$ensembl_gene_id,
                       mart = mart)

Entrez2Ensembl <- as.data.frame(org.Hs.egENSEMBL2EG)
Entrez2HGNC <- as.data.frame(org.Hs.egSYMBOL2EG)
HGNC2Ensembl <- left_join(Entrez2HGNC, Entrez2Ensembl)
HGNC2Ensembl = HGNC2Ensembl[grep('^LOC',HGNC2Ensembl$symbol),c('ensembl_id', 'symbol')] %>%
  filter(!is.na(ensembl_id), !is.na(symbol)) %>%
  dplyr::rename(ensembl_gene_id = ensembl_id, hgnc_symbol = symbol)

Ensemble2HGNC = left_join(backgroundGenes,
                          rbindlist(list(Ensemble2HGNC, HGNC2Ensembl), use.names = T, fill = T) %>%
                            unique) %>%
  unique
```

### Get differential expression results from synapse
```{r diffexp, fig.height=16, fig.width=10}
diff.exp.ids = c(ROSMAP = 'syn8018358',
                 MSBB = 'syn8073678',
                 MAYO = 'syn8028572')
diff.exp = lapply(diff.exp.ids, downloadFile)

# Filter case-control differences alone
diff.exp$ROSMAP = filter(diff.exp$ROSMAP, Comparison == 'cogdx1-cogdx4')
diff.exp$MSBB = filter(diff.exp$MSBB, Comparison == 'ND-SD')
diff.exp$MAYO = filter(diff.exp$MAYO, Comparison == 'Control-AD')

writeLines('Total number of genes differentially expressed between Control-AD at an FDR of 0.05 and foldchange of 1.2 are')
lapply(diff.exp, function(x){
  x = filter(x, adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) 
  x$Direction = 'NONE'
  x$Direction[x$logFC > 0] = 'UP'
  x$Direction[x$logFC < 0] = 'DOWN'
  x = x %>% group_by(Study, Region, Direction) %>%
    summarise(count = length(unique(gene.id)))
}) %>%
  rbindlist() %>%
  spread(Direction, count) %>%
  kable

tmp = diff.exp %>% 
  rbindlist(use.names = T, fill = T) 
tmp$Direction = 'NONE'
tmp$Direction[tmp$logFC > log2(1.2) & tmp$adj.P.Val <= 0.05] = 'UP'
tmp$Direction[tmp$logFC < -log2(1.2) & tmp$adj.P.Val <= 0.05] = 'DOWN'

pl = list()
for (study in c('ROSMAP','MSBB','MAYO')){
  p = ggplot(tmp %>% filter(Study == study) %>% droplevels, aes(x = logFC, y = -log10(adj.P.Val), color = Direction)) + geom_point()
  p = p + scale_color_manual(values = c('green','grey','red'))
  p = p + facet_grid(Study~.+Region) + theme_bw() + theme(legend.position = 'top')
  pl[[study]] = p
}
multiplot(plotlist = pl, cols = 1)

# Find genes that are differentially expressed in all brain regions
diff.exp.gl = dlply(tmp, .(Region), .fun = function(x){
  dplyr::filter(x, abs(logFC) >= log2(1.2) & adj.P.Val <= 0.05) %>%
    dplyr::select(gene.id) %>%
    unique %>% unlist
})

tmp1 = filter(tmp, gene.id %in% Reduce(intersect, diff.exp.gl)) %>%
  dplyr::select(-ensembl_gene_id, -hgnc_symbol) %>%
  left_join(Ensemble2HGNC %>% dplyr::select(-position))
tmp1$hgnc_symbol[tmp1$ensembl_gene_id == 'ENSG00000275830'] = 'LOC101927646'

tmp.eff.sz = group_by(tmp1, gene.id) %>% summarise(es = sum(logFC, na.rm = T)) %>% top_n(6, abs(es))

writeLines('Top six differentially expressed genes are')
pl = list()
p = ggplot(tmp1 %>% filter(gene.id %in% tmp.eff.sz$gene.id[tmp.eff.sz$es>0]), aes(x = Region, y = logFC, fill = Study)) + geom_bar(stat = 'identity')
p = p + facet_grid(hgnc_symbol~.) + geom_errorbar(aes(ymin=CI.L, ymax=CI.R), width = 0.2)
p = p + theme(legend.position = 'top')
pl[[1]] = p

p = ggplot(tmp1 %>% filter(gene.id %in% tmp.eff.sz$gene.id[tmp.eff.sz$es<0]), aes(x = Region, y = logFC, fill = Study)) + geom_bar(stat = 'identity')
p = p + facet_grid(hgnc_symbol~.) + geom_errorbar(aes(ymin=CI.L, ymax=CI.R), width = 0.2)
p = p + theme(legend.position = 'top')
pl[[2]] = p

multiplot(plotlist = pl, cols = 2)

# Genes differentially expressed in all brain regions
tmp1 = dlply(tmp, .(Region), .fun = function(x){
  dplyr::filter(x, abs(logFC) >= log2(1) & adj.P.Val <= 0.05) %>%
    dplyr::select(gene.id) %>%
    unique %>% unlist
}) %>%
  Reduce(intersect, .)
writeLines('Genes that are differentially expressed between Control-AD at FDR 0.05 in all brain regions')
kable(data.frame(ngenes = length(tmp1),
                 hgnc_symbol = paste(unique(Ensemble2HGNC$hgnc_symbol[Ensemble2HGNC$gene.id %in% tmp1]), collapse = ', ')))

tmp1 = dlply(tmp, .(Region), .fun = function(x){
  dplyr::filter(x, abs(logFC) >= log2(1.2) & adj.P.Val <= 0.05) %>%
    dplyr::select(gene.id) %>%
    unique %>% unlist
}) %>%
  Reduce(intersect, .)
writeLines('Genes that are differentially expressed between Control-AD at FDR 0.05 and foldchange of 1.2 in all brain regions')
kable(data.frame(ngenes = length(tmp1),
                 hgnc_symbol = paste(unique(Ensemble2HGNC$hgnc_symbol[Ensemble2HGNC$gene.id %in% tmp1]), collapse = ', ')))
```

### Meta Analysis
```{r metaanalysis.min.pval}
# Get raw p.values
P.Value = diff.exp %>% 
  rbindlist(use.names = T, fill = T) %>%
  dplyr::select(gene.id, P.Value, Region) %>%
  unique() %>%
  tidyr::spread(Region, P.Value)
P.Value[is.na(P.Value)] = 1
rownames(P.Value) = P.Value$gene.id
P.Value$gene.id = NULL
P.Value = data.matrix(P.Value)

# Perform minimum p-value based meta-analysis
min.pval = P.Value %>% 
  apply(1, min, na.rm = T) %>%
  p.adjust(method = 'bonferroni') %>%
  rownameToFirstColumn('gene.id') %>%
  dplyr::rename(min.pval = DF)

# Perform fisher p-value based meta-analysis
fb.pval = P.Value %>%
  fishersPvalue %>%
  p.adjust(method = 'bonferroni') %>%
  rownameToFirstColumn('gene.id') %>%
  dplyr::rename(fb.pval = DF)
```

```{r }
# Rank prod based analysis
expr = resid.expr[!duplicated(resid.expr$ensembl_gene_id),-c(1:2)]
rownames(expr) = resid.expr$ensembl_gene_id[!duplicated(resid.expr$ensembl_gene_id)]

cl = covariates$Status
names(cl) = covariates$ID
cl[cl == 'NCI'] = 'Control'
cl[cl == 'ND'] = 'Control'
cl[cl == 'SD'] = 'AD'
cl = cl[cl %in% c('Control','AD')] %>% droplevels 
cl.tmp = -as.numeric(cl)+2
names(cl.tmp) = names(cl)

origin = covariates$BrainRegion %>% as.numeric
names(origin) = covariates$ID
origin = origin[names(cl)]

expr = expr[, names(cl)]

rankprod.pfp = RPadvance(expr, cl.tmp, origin, num.perm = 100, gene.names = rownames(expr), plot=T, rand = 123456, huge=T)
rankprod.pval = apply(rankprod.pfp$pfp, 1, min) %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  dplyr::rename(rankprod.pval = DF)

metaAnal = join_all(list(min.pval, max.pval, fb.pval, rankprod.pval), type = 'full')
metaAnal$weight = -log10(metaAnal[,-c(1,6)]) %>% rowSums %>% rank
metaAnal = metaAnal %>% arrange(desc(weight))

writeLines('Number of genes differentially expressed between AD and control at an FDR of 0.05 are')
tmp = metaAnal %>% 
  dplyr::select(-weight) %>%
  gather(method, pval, -ensembl_gene_id) %>% 
  group_by(method) %>% 
  filter(pval <= 0.05) %>%
  dplyr::select(ensembl_gene_id) %>%
  unique %>%
  dlply(.(method), .fun = function(x){unique(x$ensembl_gene_id)})
kable(sapply(tmp, length) %>% as.data.frame)

tmp1 = Venn(tmp)

plotVenn3d(tmp1@IndicatorWeight[-(1),'.Weight'], labels = names(tmp))
```
### Heatmap of expression
```{r heatmap}
# Get the overlapping gene sets from all three methods
tmp.expr = t(scale(t(expr)))
tmp.expr = scale(tmp.expr)
tmp.expr = tmp.expr[intersect(intersect(tmp$fb.pval, tmp$min.pval), tmp$rankprod.pval),]

tmp.covariates = dplyr::select(covariates, BrainRegion, Status)
tmp.covariates$Status[tmp.covariates$Status == 'NCI'] = 'Control'
tmp.covariates$Status[tmp.covariates$Status == 'ND'] = 'Control'
tmp.covariates$Status[tmp.covariates$Status == 'SD'] = 'AD'
tmp.covariates = tmp.covariates %>% droplevels
rownames(tmp.covariates) = covariates$ID

h1 = HeatmapAnnotation(tmp.covariates)
Heatmap(tmp.expr, top_annotation = h1, show_row_names = F, show_column_names = F,
        show_row_dend = F, show_column_dend = F)
```

### Pathway enrichment analysis
```{r pathway.enrich}
setsToTest = diffexp %>%
  filter(adj.P.Val <= 0.05) %>%
  dlply(.(Tissue, Direction), .fun = function(x){
    unique(x$ensembl_gene_id)
  })

tmp1 = rankprod.pfp$pfp[intersect(intersect(tmp$fb.pval, tmp$min.pval), tmp$rankprod.pval),] <= 0.05
setsToTest$META.UP = rownames(tmp1)[which(tmp1[,1])]
setsToTest$META.DOWN = rownames(tmp1)[which(tmp1[,2])]

# Convert ensembl id to hgnc_symbols
setsToTest = lapply(setsToTest, function(x, Ensemble2HGNC){
  Ensemble2HGNC$hgnc_symbol[Ensemble2HGNC$ensembl_gene_id %in% x] %>%
    unique()
}, Ensemble2HGNC)

# Get Background genes
backgroundGenes = Ensemble2HGNC$hgnc_symbol[Ensemble2HGNC$ensembl_gene_id %in% rownames(expr)] %>% unique()

############################################################################################################
# Function to perform Fishers enrichment analysis
fisherEnrichment <- function(genesInSignificantSet, # A character vector of differentially expressed or some significant genes to test
                             genesInGeneSet, # A character vector of genes in gene set like GO annotations, pathways etc...
                             genesInBackground # Background genes that are 
){
  genesInSignificantSet = intersect(genesInSignificantSet, genesInBackground) # back ground filtering
  genesInNonSignificantSet = base::setdiff(genesInBackground, genesInSignificantSet)
  genesInGeneSet = intersect(genesInGeneSet, genesInBackground) # back ground filtering
  genesOutGeneSet = base::setdiff(genesInBackground,genesInGeneSet)
  
  pval = fisher.test(
    matrix(c(length(intersect(genesInGeneSet, genesInSignificantSet)),             
             length(intersect(genesInGeneSet, genesInNonSignificantSet)),
             length(intersect(genesOutGeneSet, genesInSignificantSet)),
             length(intersect(genesOutGeneSet, genesInNonSignificantSet))), 
           nrow=2, ncol=2), alternative="greater")
  OR = (length(intersect(genesInGeneSet, genesInSignificantSet)) * length(intersect(genesOutGeneSet, genesInNonSignificantSet))) / (length(intersect(genesInGeneSet, genesInNonSignificantSet)) * length(intersect(genesOutGeneSet, genesInSignificantSet))) %>% as.numeric()
  return(data.frame(pval = pval$p.value,
                    ngenes = length(genesInGeneSet),
                    noverlap = length(intersect(genesInGeneSet, genesInSignificantSet)),
                    Odds.Ratio = OR,
                    Genes = paste(intersect(genesInGeneSet, genesInSignificantSet), collapse = '|')))}
############################################################################################################

############################################################################################################
#### Get gene sets ####
# Download enrichr gene sets from synapse
GL_OBJ = synGet('syn4867851')
ALL_USED_IDs = c(ALL_USED_IDs, GL_OBJ$properties$id)
load(GL_OBJ@filePath)

gsets = c("BioCarta_2015", "GO_Biological_Process", "Reactome_2015", "TargetScan_microRNA", "WikiPathways_2015")
GeneSets.Enrichr = GeneSets[gsets]

# Download AD related gene sets from synapse
GL_OBJ = synGet('syn4893059');
ALL_USED_IDs = c(ALL_USED_IDs, GL_OBJ$properties$id)
load(GL_OBJ@filePath)

GeneSets.CM = list(CellTypeMarkers = GeneSets[grep('Zhang', names(GeneSets))],
                   AD = GeneSets['AD:GeneticLoci'])
GeneSets = c(GeneSets.Enrichr, GeneSets.CM)

# Filter gene list
GeneSets = filterGeneSets(GeneSets, backgroundGenes, minSize = 3, maxSize = 5000)

enrichResults = list()
for (name in names(setsToTest)){
  genesInModule = setsToTest[[name]]  
  enrichResults[[name]] = lapply(GeneSets,
                                    function(x, genesInModule, genesInBackground, fisherEnrichment, rownameToFirstColumn){
                                      tmp = as.data.frame(t(sapply(x, fisherEnrichment, genesInModule, genesInBackground)))
                                      tmp = rownameToFirstColumn(tmp,'GeneSetName')
                                      return(tmp)
                                    }, unique(genesInModule), unique(backgroundGenes), fisherEnrichment, rownameToFirstColumn) %>%
    rbindlist(use.names=TRUE, idcol = 'Category') %>%
    dplyr::mutate(fdr = p.adjust(pval, 'fdr'))
  writeLines(paste0('Completed ',name))  
}

# Write results to file
tmp1 = rbindlist(enrichResults, use.names = TRUE, idcol = 'ComparisonName', fill = TRUE) %>%
  filter(!is.na(Category))

tmp1$pval = unlist(tmp1$pval)
tmp1$ngenes = unlist(tmp1$ngenes)
tmp1$noverlap = unlist(tmp1$noverlap)
tmp1$Odds.Ratio = unlist(tmp1$Odds.Ratio)
tmp1$Genes = unlist(tmp1$Genes)

# Store logcpm in synapse
write.table(tmp1, file = 'enrich.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('enrich.tsv', 
           name = 'Pathway Enrichment Results (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')
```

### Store results in synapse
```{r synapse.store, include=FALSE, cache=FALSE}
thisFileName <- 'MetaAnalysis.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='metaAnal')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))

# Store consolidated data in synapse
CODE = Folder(name = 'MetaAnalysis of Control vs AD Data', parentId = "syn5569102")
CODE = synStore(CODE)

# Store covariates in synapse
write.table(covariates, file = 'ALLCovariates.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLCovariates.tsv', 
           name = 'Covariates', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')

# Store logcpm in synapse
write.table(logcpm, file = 'ALLLogCPM.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLLogCPM.tsv', 
           name = 'Log CPM (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')

# Store residual expression in synapse
write.table(expr, file = 'ALLExpr.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLExpr.tsv', 
           name = 'Residual Expression (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')

# Store diffexp in synapse
write.table(diffexp, file = 'ALLDiffExp.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLDiffExp.tsv', 
           name = 'Differential Expression (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')

# Store meta analysis in synapse
write.table(metaAnal, file = 'ALLMetaAnal.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('ALLMetaAnal.tsv', 
           name = 'Meta Analysis (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')
```