---
title: "Meta analysis of differential expression results from AMP-AD (Control vs AD, reprocessed RNASeq)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knit2synapse)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./MetaAnalysis.Rmd", 
                                 parentId ="syn9622393",
                                 entityName = 'MetaAnalysis of Control vs AD (Reprocessed RNASeq)')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library(CovariateAnalysis) # install from th1vairam repo using devtools::install_github('th1vairam/CovariateAnalysis@devtools')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(RColorBrewer)
library(ggplot2)
library(ComplexHeatmap)
library(biomaRt)
library(Matrix)
library(org.Hs.eg.db)

library(limma)
library(edgeR)
library(RankProd)
library(githubr)

library(Vennerable)
library(colorfulVennPlot)
library(rmeta)
library(xlsx)

synapseLogin()

library(doParallel)
library(foreach)
library(doParallel)

clust = makeCluster(detectCores() - 2)
registerDoParallel(clust)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r fxns}
# Utility function for combining p-values using Fishers inverse chi-square method
fishersPvalue <- function(pvalue){
  pvalue = -2*log(pvalue, base = exp(1))
  S <- 0
  N <- dim(pvalue)[2]
  
  correlation <- WGCNA::bicor(pvalue,use = 'pairwise.complete.obs');
  
  for (i in 1:(N-1))
    for (j in (i+1):N){
      if (correlation[i,j] > 0){
        S <- S + correlation[i,j]*(3.25+0.75*correlation[i,j])
      } else if (correlation[i,j] <= 0){
        S <- S + correlation[i,j]*(3.27+0.71*correlation[i,j])
      }
    }
  
  combinedPvalue.chi <- rowSums(pvalue)*(4*N)/(4*N+2*S)
  combinedPvalue <- sapply(combinedPvalue.chi, function(x,N){
    pchisq(x, df = 2*N, lower.tail = FALSE)
  },N)
  names(combinedPvalue) = rownames(pvalue)
  
  return(combinedPvalue)
}
```

### Get curated gene list
```{r curate.genelist}
# Download enrichr gene sets from synapse
GL_OBJ = synGet('syn4867851')
all.used.ids = 'syn4867851'
load(GL_OBJ@filePath)

gsets = c("GO_Biological_Process_2015", "Reactome_2016","KEGG_2016")
GeneSets.Enrichr = GeneSets[gsets]

# Download AD genetic loci from synapse
AD = list(GeneticLoci = synGet('syn4891675')@filePath %>%
            read.table(header = F) %>% 
            unlist() %>%
            as.character())
all.used.ids = c(all.used.ids, 'syn4891675')

# Download cell type markers from synapse (list from GSE73721 by Mariet)
load(synGet('syn9818080')@filePath)
all.used.ids = c(all.used.ids, 'syn9818080')

GeneSets = c(list(CellTypeMarkers = CellTypeList,
                AD = AD),
                GeneSets.Enrichr)
```

### Get covariates/sample metadata from synapse
```{r covariates}
covariates = downloadFile('syn9765937')
all.used.ids = c(all.used.ids, 'syn9765937')

kable(table(covariates$Diagnosis, covariates$BrainRegion))
```

### Get differential expression data from synapse
```{r diffexp, fig.width=10, fig.height=15}
diff.exp.ids = c(ROSMAP = 'syn8456721',
                 MSBB = 'syn8502894',
                 MAYO = 'syn8468023')
all.used.ids = c(all.used.ids, diff.exp.ids)
diff.exp = lapply(diff.exp.ids, downloadFile)

diff.exp$ROSMAP = diff.exp$ROSMAP %>%
  setnames(c('Region'), c('BrainRegion'))

tmp = rbindlist(diff.exp, use.names = T, fill = T) %>%
  dplyr::filter(Comparison == 'CONTROL-AD')

tmp$Direction = 'NONE'
tmp$Direction[tmp$logFC >= log2(1.2) & tmp$adj.P.Val <= 0.05] = 'UP'
tmp$Direction[tmp$logFC <= -log2(1.2) & tmp$adj.P.Val <= 0.05] = 'DOWN'

writeLines('Number of differentially expressed genes at an FDR of 0.05 and fold change of 1.2')
tmp %>%
  group_by(Study, BrainRegion, Direction, Comparison) %>%
  summarise(count = n()) %>%
  spread(Direction, count) %>%
  kable

pl = list()
for(region in c('ROSMAP', 'MSSM', 'MAYO')){
  p = ggplot(tmp %>% filter(Study == region), aes(x = logFC, y = -log10(adj.P.Val), color = Direction)) + geom_point()
  p = p + scale_color_manual(values = c('green','grey','red'))
  p = p + facet_grid(Study~.+BrainRegion) + theme(legend.position = 'top')
  pl[[region]] = p
}
multiplot(plotlist = pl, cols = 1)

diffExpSet = plyr::dlply(tmp, .(BrainRegion), .fun = function(x){
  dplyr::filter(x, abs(logFC) >= log2(1.2), adj.P.Val <= 0.05) %>%
    dplyr::select(hgnc_symbol) %>%
    unlist %>%
    as.character() %>%
    unique() %>%
    setdiff('')
})
```

### Meta Analysis
```{r metaanalysis.min.pval}
library(metamisc)
metaAnal <- function(diff.exp, region){
  # Get raw p.values
  P.Value = diff.exp %>% 
    rbindlist(use.names = T, fill = T) %>%
    dplyr::filter(Comparison == 'CONTROL-AD', BrainRegion %in% region) %>%
    dplyr::select(ensembl_gene_id, P.Value, BrainRegion) %>%
    unique() %>%
    tidyr::spread(BrainRegion, P.Value)
  P.Value[is.na(P.Value)] = 1
  rownames(P.Value) = P.Value$ensembl_gene_id
  P.Value$ensembl_gene_id = NULL
  P.Value = data.matrix(P.Value)

  # Perform fisher p-value based meta-analysis
  fb.pval = P.Value %>%
    fishersPvalue %>%
    p.adjust(method = 'fdr') %>%
    rownameToFirstColumn('gene.id') %>%
    dplyr::rename(fb.pval = DF)
  sum(fb.pval$fb.pval <= 0.05)

  # Dersimonian based meta analysis
  effect.size = diff.exp %>% 
    rbindlist(use.names = T, fill = T) %>%
    dplyr::filter(Comparison == 'CONTROL-AD', BrainRegion %in% region) %>%
    dplyr::select(ensembl_gene_id, logFC, BrainRegion) %>%
    unique() %>%
    tidyr::spread(BrainRegion, logFC)
  rownames(effect.size) = effect.size$ensembl_gene_id
  effect.size$ensembl_gene_id = NULL
  effect.size[is.na(effect.size)] = 0

  effect.var = diff.exp %>% 
    rbindlist(use.names = T, fill = T) %>%
    dplyr::filter(Comparison == 'CONTROL-AD', BrainRegion %in% region) %>%
    dplyr::mutate(sd = (CI.R - logFC)/3,
                  var = sd^2) %>%
    dplyr::select(ensembl_gene_id, var, BrainRegion) %>%
    unique() %>%
    tidyr::spread(BrainRegion, var)
  rownames(effect.var) = effect.var$ensembl_gene_id
  effect.var$ensembl_gene_id = NULL
  effect.var[is.na(effect.var)] = 0

  ds.metaAnal = ldply(1:dim(effect.size)[1], function(i, effect.size, effect.var){
    es = t(effect.size[i,]); es = es[es!=0]
    ev = t(effect.var[i,]); ev = ev[ev!=0]
  fit = uvmeta(es, ev, model = "random", method = "MOM")
  data.frame(ensembl_gene_id = rownames(effect.size)[i],
             logFC = fit$results[1,1],
             CI.L = fit$results[1,1] - 3 * sqrt(fit$results[1,2]),
             CI.R = fit$results[1,1] + 3 * sqrt(fit$results[1,2]))
  }, effect.size, effect.var)

  tmp = dplyr::rename(fb.pval, ensembl_gene_id = gene.id) %>%
    left_join(ds.metaAnal) %>%
    dplyr::filter(fb.pval <= 0.05, sign(CI.L) == sign(CI.R), abs(logFC) >= log2(1.2)) %>%
    dplyr::arrange(desc(logFC)) %>%
    left_join(diff.exp %>%
                rbindlist(use.names = T, fill = T) %>%
                dplyr::select(ensembl_gene_id, hgnc_symbol) %>%
                unique())
}

meta.anal.all= metaAnal(diff.exp, c('DLPFC', 'CER', 'TCX', 'FP', 'STG', 'PHG', 'IFG'))
meta.anal.imp= metaAnal(diff.exp, c('DLPFC', 'TCX', 'PHG'))
diffExpSet = c(diffExpSet, list(METANLZ.ALL = setdiff(meta.anal.all$hgnc_symbol, ''),
                                METANLZ.T3 = setdiff(meta.anal.imp$hgnc_symbol, '')))
```

### Pathway enrichment analysis (individual brain regions)
```{r pathway.enrich}
backgroundGenes = diff.exp %>%
  rbindlist(use.names = T, fill = T) %>%
  dplyr::select(hgnc_symbol) %>%
  unlist() %>%
  setdiff('')

# Filter geneset
GeneSets = lapply(GeneSets, function(x, backgroundGenes){
  x = intersect(x,backgroundGenes)
}, backgroundGenes)
GeneSets = GeneSets[sapply(GeneSets, length) >= 15 & sapply(GeneSets, length) <= 1000]

enrichResults = plyr::ldply(diffExpSet, .fun = function(genesInGeneSet, geneSets, backgroundGenes){
  plyr::ldply(geneSets, .fun = fisherEnrichment, genesInGeneSet, backgroundGenes, .parallel = T, .id = 'Category') %>%
    dplyr::mutate(fdr = p.adjust(pval, method = 'fdr'))
}, GeneSets, backgroundGenes, 
.parallel = F, .progress = 'text', .paropts = list(.packages = c('plyr','dplyr','CovariateAnalysis')),
.id = 'BrainRegion')

tmp = dplyr::select(enrichResults, BrainRegion, Category, fdr, Odds.Ratio)
tmp = tmp[c(grep('CellTypeMarkers', tmp$Category),
            grep('GeneticLoci', tmp$Category)), ]
tmp$Odds.Ratio[tmp$fdr > 0.05] = 0
tmp = tmp %>%
  dplyr::select(BrainRegion, Category, Odds.Ratio) %>%
  tidyr::spread(BrainRegion, Odds.Ratio)
rownames(tmp) = gsub('CellTypeMarkers.','',tmp$Category) %>%
  gsub('_Zhang_Neuron_2015','',.)
tmp$Category = NULL

h = ComplexHeatmap::Heatmap(tmp, name = 'Odds Ratio', col = wesanderson::wes_palette('Zissou', 4))
draw(h, heatmap_legend_side = 'left', padding = unit(c(4,2,2,4), 'mm'))

tmp = dplyr::select(enrichResults, BrainRegion, Category, fdr, Odds.Ratio)
tmp = tmp[c(grep('Reactome', tmp$Category),
            grep('GO', tmp$Category)), ] %>%
  unique()
tmp$Odds.Ratio[tmp$fdr > 0.05] = 0
tmp = tmp %>%
  dplyr::select(BrainRegion, Category, Odds.Ratio) %>%
  tidyr::spread(BrainRegion, Odds.Ratio)
tmp = tmp[rowSums(tmp[,-(1)]) != 0, c(1, which(colSums(tmp[,-(1)]) != 0) + 1)]
rownames(tmp) = gsub('GO_Biological_Process_2015.','',tmp$Category) %>%
  gsub('Reactome_2016.','',.)
tmp$Category = NULL

h = ComplexHeatmap::Heatmap(tmp, name = 'Odds Ratio', col = wesanderson::wes_palette('Zissou', 4))
draw(h, heatmap_legend_side = 'left', padding = unit(c(4,2,2,4), 'mm'))
```

### Gender specific differential expression (individual brain regions)
```{r gender.specific}
load(synGet('syn9766351')@filePath)

tmp = ldply(all.results, .fun = function(x){
  x$Direction = 'NONE'
  x$Direction[x$adj.P.Val <= 0.05 & x$logFC <= -log2(1.2)] = 'DOWN'
  x$Direction[x$adj.P.Val <= 0.05 & x$logFC >= log2(1.2)] = 'UP'
  
  dplyr::filter(x, adj.P.Val <= 0.05, abs(logFC) >= log2(1.2), Comparison == 'AD.FEMALE-CONTROL.FEMALE') %>%
    dplyr::group_by(Direction) %>%
    dplyr::summarise(count = n())
}) %>%
  dplyr::mutate(Direction = paste(Direction,'FEMALE',sep = '.')) %>%
  tidyr::spread(Direction, count) %>%
  left_join(
    ldply(all.results, .fun = function(x){
      x$Direction = 'NONE'
      x$Direction[x$adj.P.Val <= 0.05 & x$logFC <= -log2(1.2)] = 'DOWN'
      x$Direction[x$adj.P.Val <= 0.05 & x$logFC >= log2(1.2)] = 'UP'
      
      dplyr::filter(x, adj.P.Val <= 0.05, abs(logFC) >= log2(1.2), Comparison == 'AD.MALE-CONTROL.MALE') %>%
        dplyr::group_by(Direction) %>%
        dplyr::summarise(count = n())
    }) %>%
      dplyr::mutate(Direction = paste(Direction,'MALE',sep = '.')) %>%
      tidyr::spread(Direction, count))

metaAnal.gender <- function(allResults, comp, region){
  # Get raw p.values
  P.Value = allResults %>%
    rbindlist(use.names = T, fill = T, idcol = 'BrainRegion') %>%
    dplyr::filter(Comparison == comp, BrainRegion %in% region) %>%
    dplyr::select(ensembl_gene_id, P.Value, BrainRegion) %>%
    unique() %>%
    tidyr::spread(BrainRegion, P.Value)
  P.Value[is.na(P.Value)] = 1
  rownames(P.Value) = P.Value$ensembl_gene_id
  P.Value$ensembl_gene_id = NULL
  P.Value = data.matrix(P.Value)
  
  # Perform fisher p-value based meta-analysis
  fb.pval = P.Value %>%
    fishersPvalue %>%
    p.adjust(method = 'fdr') %>%
    rownameToFirstColumn('gene.id') %>%
    dplyr::rename(fb.pval = DF)
  sum(fb.pval$fb.pval <= 0.05)
  
  # Dersimonian based meta analysis
  effect.size = allResults %>% 
    rbindlist(use.names = T, fill = T, idcol = 'BrainRegion') %>%
    dplyr::filter(Comparison == comp, BrainRegion %in% region) %>%
    dplyr::select(ensembl_gene_id, logFC, BrainRegion) %>%
    unique() %>%
    tidyr::spread(BrainRegion, logFC)
  rownames(effect.size) = effect.size$ensembl_gene_id
  effect.size$ensembl_gene_id = NULL
  effect.size[is.na(effect.size)] = 0
  
  effect.var = allResults %>% 
    rbindlist(use.names = T, fill = T, idcol = 'BrainRegion') %>%
    dplyr::filter(Comparison == comp, BrainRegion %in% region) %>%
    dplyr::mutate(sd = (CI.R - logFC)/3,
                  var = sd^2) %>%
    dplyr::select(ensembl_gene_id, var, BrainRegion) %>%
    unique() %>%
    tidyr::spread(BrainRegion, var)
  rownames(effect.var) = effect.var$ensembl_gene_id
  effect.var$ensembl_gene_id = NULL
  effect.var[is.na(effect.var)] = 0
  
  ds.metaAnal = ldply(1:dim(effect.size)[1], function(i, effect.size, effect.var){
    es = t(effect.size[i,]); 
    ev = t(effect.var[i,]); 
    es1 = es[es != 0 & ev != 0]; ev1 = ev[ev != 0 & es != 0]
    fit = uvmeta(es1, ev1, model = "random", method = "MOM")
    data.frame(ensembl_gene_id = rownames(effect.size)[i],
               logFC = fit$results[1,1],
               CI.L = fit$results[1,1] - 3 * sqrt(fit$results[1,2]),
               CI.R = fit$results[1,1] + 3 * sqrt(fit$results[1,2]))
  }, effect.size, effect.var)
  
  tmp = dplyr::rename(fb.pval, ensembl_gene_id = gene.id) %>%
    left_join(ds.metaAnal) %>%
    dplyr::filter(fb.pval <= 0.05, sign(CI.L) == sign(CI.R)) %>%
    dplyr::arrange(desc(logFC)) %>%
    left_join(diff.exp %>%
                rbindlist(use.names = T, fill = T) %>%
                dplyr::select(ensembl_gene_id, hgnc_symbol) %>%
                unique())
}

meta.female = metaAnal.gender(all.results, comp = 'AD.FEMALE-CONTROL.FEMALE', 
                              region = c('DLPFC', 'CER', 'TCX', 'FP', 'STG', 'PHG', 'IFG'))
meta.male = metaAnal.gender(all.results, comp = 'AD.MALE-CONTROL.MALE',
                            region = c('DLPFC', 'CER', 'TCX', 'FP', 'STG', 'PHG', 'IFG'))
meta.female.imp = metaAnal.gender(all.results, comp = 'AD.FEMALE-CONTROL.FEMALE', 
                                  region = c('DLPFC', 'TCX', 'PHG'))
meta.male.imp = metaAnal.gender(all.results, comp = 'AD.MALE-CONTROL.MALE',
                                region = c('DLPFC', 'TCX', 'PHG'))

diffExpSet.gender = llply(all.results, .fun = function(x){
  dplyr::filter(x, adj.P.Val <= 0.05, abs(logFC) >= log2(1.2), Comparison == 'AD.FEMALE-CONTROL.FEMALE') %>%
    dplyr::left_join(diff.exp %>%
                       rbindlist(use.names = T, fill = T) %>%
                       dplyr::select(ensembl_gene_id, hgnc_symbol) %>%
                       unique) %>%
    dplyr::select(hgnc_symbol) %>%
    unlist() %>% setdiff('')
})

diffExpSet.gender = c(diffExpSet.gender, llply(all.results, .fun = function(x){
  dplyr::filter(x, adj.P.Val <= 0.05, abs(logFC) >= log2(1.2), Comparison == 'AD.MALE-CONTROL.MALE') %>%
    dplyr::left_join(diff.exp %>%
                       rbindlist(use.names = T, fill = T) %>%
                       dplyr::select(ensembl_gene_id, hgnc_symbol) %>%
                       unique) %>%
    dplyr::select(hgnc_symbol) %>%
    unlist() %>% setdiff('')
}))
names(diffExpSet.gender)[1:7] = paste('FEMALE',names(diffExpSet.gender)[1:7],sep = '.')
names(diffExpSet.gender)[8:14] = paste('MALE',names(diffExpSet.gender)[8:14],sep = '.')
diffExpSet.gender$FEMALE.METANALZ.ALL = setdiff(meta.female$hgnc_symbol,'')
diffExpSet.gender$MALE.METANALZ.ALL = setdiff(meta.male$hgnc_symbol,'')
diffExpSet.gender$FEMALE.METANALZ.T3 = setdiff(meta.female.imp$hgnc_symbol,'')
diffExpSet.gender$MALE.METANALZ.T3 = setdiff(meta.male.imp$hgnc_symbol,'')

enrichResults.gender = plyr::ldply(diffExpSet.gender, .fun = function(genesInGeneSet, geneSets, backgroundGenes){
  plyr::ldply(geneSets, .fun = fisherEnrichment, genesInGeneSet, backgroundGenes, .parallel = T, .id = 'Category') %>%
    dplyr::mutate(fdr = p.adjust(pval, method = 'fdr'))
}, GeneSets, backgroundGenes, 
.parallel = F, .progress = 'text', .paropts = list(.packages = c('plyr','dplyr','CovariateAnalysis')),
.id = 'BrainRegion')

tmp = dplyr::select(enrichResults.gender, BrainRegion, Category, fdr, Odds.Ratio)
tmp = tmp[c(grep('CellTypeMarkers', tmp$Category),
            grep('GeneticLoci', tmp$Category)), ]
tmp$Odds.Ratio[tmp$fdr > 0.05] = 0
tmp = tmp %>%
  dplyr::select(BrainRegion, Category, Odds.Ratio) %>%
  tidyr::spread(BrainRegion, Odds.Ratio)
rownames(tmp) = gsub('CellTypeMarkers.','',tmp$Category) %>%
  gsub('_Zhang_Neuron_2015','',.)
tmp$Category = NULL

h = ComplexHeatmap::Heatmap(t(tmp), name = 'Odds Ratio', col = wesanderson::wes_palette('Zissou', 4))
draw(h, heatmap_legend_side = 'left', padding = unit(c(4,2,2,10), 'mm'))

```

### Store results in synapse
```{r synapse.store, include=FALSE, cache=FALSE}
thisFileName <- 'MetaAnalysis.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='metaAnal')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))

# Store consolidated data in synapse
CODE = Folder(name = 'MetaAnalysis of Control vs AD Data', parentId = "syn5569102")
CODE = synStore(CODE)

# Store logcpm in synapse
write.table(tmp, file = 'enrich.tsv', sep = '\t', row.names=F, quote=F)
OBJ = File('enrich.tsv', 
           name = 'Pathway Enrichment Results (AD and Control)', 
           parentId = CODE$properties$id)
OBJ = synStore(OBJ,
               used = as.character(c(covariates.id)),
               executed = thisFile,
               activityName = 'MetaAnalysis of AMP-AD reprocessed data')
```