---
title: "Markdown to estimate the abundance of brain cell types with the reprocessed counts from MSSM cohorts (with TMM-VOOM normalisation)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knit2synapse)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./MSSM_RNASeq_Reprocessed_TMM_VOOM_CellProp.Rmd",
                                 parentId = "syn5570325",
                                 entityName = 'MSSM Cell Type Abundance with Reprocessed RNASeq')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## It is assumed your working directory is where this file

## Load required libraries
library(CovariateAnalysis) # get it from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(reshape2)
library(limma)
library(Biobase)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(RColorBrewer)
library(xlsx)

library(synapseClient)
library(knitr)
library(githubr)

library(parallel)
library(doParallel)
library(doMC)
library(foreach)

cl = makeCluster(detectCores() - 2)
registerDoParallel(cl)

synapseLogin()

# Source external files
source('../R/CIBERSORT.R')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn5570325';
activityName = 'Cell type abundance in MSSM';
activityDescription = 'Cell type abundance in MSSM with TMM-VOOM normalised reprocessed RNASeq counts';

thisFileName <- 'MSSM_RNASeq_Reprocessed_TMM_VOOM_CellProp.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='celltype')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
### Data download
Data were obtained from AMP-AD portal of MSSM, and following brain regions were analysed for differential expression between case and control:
  | Region | Name                           | 
  | :----- | ------------------------------ | 
  | BM_10  | Frontal Pole (FP)              |
  | BM_22  | Superior Temporal Gyrus (STG)  |
  | BM_36  | Parahipocampal Gyrus (PHG)     |
  | BM_44  | inferior frontal Gyrus (IFG)   |
  
* BM 10: frontopolar prefrontal cortex, rostrolateral prefrontal cortex, or anterior prefrontal cortex, 
* BM 22: In humans it corresponds approximately to the lateral and caudal two thirds of the superior temporal gyrus, 
* BM 36: Ectorhinal area 36 is a subdivision of the cytoarchitecturally defined temporal region of cerebral cortex,
* BM 44: pars opercularis (of the inferior frontal gyrus), and it refers to a subdivision of the frontal region of cerebral cortex.

#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Get reprocessed sample counts for temporal cortex
COUNT_ID <- 'syn7994853';
ALL_USED_IDs <- COUNT_ID
COUNT <- read.table(synGet(COUNT_ID)@filePath, header=T, sep='\t', check.names = F)

# Get sample ids
SampleID = data.frame(SampleID = colnames(COUNT),
                 ID = colnames(COUNT)) %>% 
  tidyr::separate(ID, c('A','B','ID'), sep = '_') 

SampleID = SampleID %>%
  dplyr::select(SampleID, ID) %>%
  dplyr::mutate(ID = as.numeric(ID))

# Get technical metadata
METADATA_TECH_ID <- 'syn6100548'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_TECH_ID
METADATA_TECH <- read.table(synGet(METADATA_TECH_ID)@filePath,sep=',',header=T, row.names=1) %>%
  group_by(sampleIdentifier) %>%
  top_n(1, -rRNA.rate) %>%
  dplyr::select(sampleIdentifier, BrodmannArea, barcode, individualIdentifier, batch, RIN, TotalReads, Mapped, rRNA.rate) %>%
  unique

# Get clinical metadata
METADATA_CLINICAL_ID <- 'syn6101474'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_CLINICAL_ID
METADATA_CLINICAL <- read.csv(synGet(METADATA_CLINICAL_ID)@filePath, header = T)

METADATA <- full_join(METADATA_TECH, METADATA_CLINICAL)
```

### Data preprocessing
```{r preprocess.data,cache=FALSE, echo=TRUE}
# Fix metadata
METADATA$AOD = gsub('\\+','',METADATA$AOD)
METADATA$Dx[METADATA$CDR <= 1] = 'ND'
METADATA$Dx[METADATA$CDR == 2 | METADATA$CDR == 3] = 'MD'
METADATA$Dx[METADATA$CDR >= 4] = 'SD'
METADATA$BrodmannArea.Dx = paste(METADATA$BrodmannArea, METADATA$Dx, sep = '.')
METADATA$RIN2 = METADATA$RIN^2
levels(METADATA$batch)[levels(METADATA$batch) == ''] = 'NoBatch'

METADATA <- METADATA %>%
  filter(!is.na(BrodmannArea), !is.na(PMI), !is.na(RIN))
```

```{r preprocess.data1}
# Match covariates to expression data
indToRetain = intersect(METADATA$sampleIdentifier, colnames(COUNT))
indRemoved = setdiff(colnames(COUNT), METADATA$sampleIdentifier)

COUNT <- COUNT[, indToRetain]

METADATA = as.data.frame(METADATA)
rownames(METADATA) = METADATA$sampleIdentifier
METADATA = METADATA[indToRetain,]
```
`r dim(COUNT)[2]` samples from `r length(unique(METADATA$individualIdentifier))` subjects were obtained from the MSBB cohorts in AMP-AD reprocessed rnaseq project 

Following `r length(indRemoved)` samples are removed:`r paste(indRemoved, collapse = ', ')`
### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering}
FactorCovariates <- c("individualIdentifier", "batch", "RACE", "SEX", "BrodmannArea.Dx")
ContCovariates <- c("RIN", "RIN2", "AOD","PMI", "TotalReads", "Mapped", "rRNA.rate")
  
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- METADATA$sampleIdentifier

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
### Gene Filtering
Preprocess counts matrix and metadata. Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of samples per BrodmannArea.Dx
```{r cpmnormalisation}
genesToAnalyze = COVARIATES %>%
  rownameToFirstColumn('sampleIdentifier') %>%
  dlply(.(BrodmannArea.Dx), .fun = function(mtd, count){
    processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$sampleIdentifier],
                                                     MIN_GENE_CPM=1, 
                                                     MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
    processed.counts$filteredExprMatrix$genes
    }, COUNT)
genesToAnalyze = unlist(genesToAnalyze) %>% unique()

PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, ], MIN_GENE_CPM=0, MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples
### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)
```{r initial.voom.normalisation, fig.height=5, fig.width=5}
# TMM normalisation
TMM.GENE_EXPRESSION = calcNormFactors(PROCESSED_COUNTS$filteredExprMatrix, method = 'TMM')

# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(TMM.GENE_EXPRESSION, design=NULL, plot=T)

# Set gene counts in specific samples that are deviating 3 sd from other samples to NA
log.mat = apply(VOOM.GENE_EXPRESSION$E, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  return((x < (mn-3*std.dev)) | (x > (mn+3*std.dev)))
}) %>% t
PROCESSED_COUNTS$filteredExprMatrix[log.mat] = NA

# TMM normalisation
TMM.GENE_EXPRESSION = calcNormFactors(PROCESSED_COUNTS$filteredExprMatrix, method = 'TMM')

# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(TMM.GENE_EXPRESSION, design=NULL, plot=T)
```
### Get cell type specific signatures and expression data from synapse
```{r get.signatures}
ALL_USED_IDs = c(ALL_USED_IDs, 'syn8068511', 'syn8068475')
# Signatures from Zhang,Y et al., 2014
sig.zhang = downloadFile('syn8068511') %>%
  dlply(.(Set), .fun = function(x){return(unique(x$Gene))})

# Signatures from Darmanis, S et al., 2015
sig.dar.obj = synGet('syn8068475')  
sig.dar.unbiased = read.xlsx(sig.dar.obj@filePath, sheetName = 'Unbiased') %>%
  lapply(function(x){as.character(unique(x))})
sig.dar.unbiased = sig.dar.unbiased[1:10]

# Extract all signatures
sig = list()
sig$OPC = unique(c(sig.zhang$OPC, sig.dar.unbiased$Cluster.1))
sig$Oligodendrocyte = unique(c(sig.zhang$MyelinOligos, sig.zhang$NewOligos, sig.dar.unbiased$Cluster.3))
sig$Neuron = unique(c(sig.zhang$Neuron, sig.dar.unbiased$Cluster.6, sig.dar.unbiased$Cluster.7,
                      sig.dar.unbiased$Cluster.10, sig.dar.unbiased$Cluster.2))
sig$Endothelial = unique(c(sig.zhang$Endothelial, sig.dar.unbiased$Cluster.8))
sig$Astrocyte = unique(c(sig.zhang$Astrocyte, sig.dar.unbiased$Cluster.4))
sig$Microglia = unique(c(sig.zhang$Astrocyte, sig.dar.unbiased$Cluster.5))

sig = ldply(sig, .fun = function(x){as.data.frame(x)})
colnames(sig) = c('CellType','Gene.Name')

# Get reference expression data from Darmanis., S et al., 2015
ALL_USED_IDs = c(ALL_USED_IDs, 'syn8077142', 'syn8077191')
ref.expr = downloadFile('syn8077142')
rownames(ref.expr) = ref.expr$hgnc_symbol
ref.expr$hgnc_symbol = NULL

ref.cellTypes = downloadFile('syn8077191') %>%
  dplyr::select(SampleID, CellType)

ref.expr1 = list()
for(celltype in unique(ref.cellTypes$CellType)){
  ids = ref.cellTypes$SampleID[ref.cellTypes$CellType == celltype]
  ref.expr1[[celltype]] = log2(rowSums(2^ref.expr[,ids], na.rm = T)) %>%
    rownameToFirstColumn('hgnc_symbol') %>%
    plyr::rename(c('DF' = celltype))
}
ref.expr1 = join_all(ref.expr1)
```

```{r get.gene.symbols, include=FALSE}
backgroundGenes = data.frame(gene.id = rownames(VOOM.GENE_EXPRESSION$E)) %>%
  dplyr::mutate(ensembl_gene_id = gene.id) %>%
  tidyr::separate(ensembl_gene_id, c('ensembl_gene_id', 'position'), sep = '\\.')

# Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "dec2016.archive.ensembl.org", dataset = "hsapiens_gene_ensembl")

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", 
                       values = backgroundGenes$ensembl_gene_id,
                       mart = mart)

# Add gene names
EXPR <- VOOM.GENE_EXPRESSION$E %>%
  rownameToFirstColumn('gene.id') %>%
  left_join(backgroundGenes) %>%
  left_join(Ensemble2HGNC) %>%
  filter(!duplicated(gene.id))
rownames(EXPR) = EXPR$gene.id

EXPR1 = WGCNA::collapseRows(EXPR %>% dplyr::select(-ensembl_gene_id, -position, -hgnc_symbol, -gene.id), 
                            EXPR$hgnc_symbol, 
                            EXPR$gene.id)$datETcollapsed

ind = intersect(rownames(EXPR1), ref.expr1$hgnc_symbol)
EXPR1 = EXPR1 %>% 
  rownameToFirstColumn('hgnc_symbol') %>%
  filter(hgnc_symbol %in% ind)
ref.expr1 = ref.expr1 %>%
  filter(hgnc_symbol %in% ind)

write.table(ref.expr1, file = 'ref.expr.txt', sep = '\t', quote=F, row.names = F)
write.table(EXPR1, file = 'mixture.expr.txt', sep = '\t', quote=F, row.names = F)

cellprop = CIBERSORT('ref.expr.txt', 'mixture.expr.txt', perm = 100) # Obtain CIBERSORT.R function from http://cibersort.stanford.edu/
stopCluster(cl)
```

```{r plot.cellprop, fig.width=16, fig.height=8}
cellprop1 = as.data.frame(cellprop)
cellprop1$BrodmannArea.Dx = COVARIATES[rownames(cellprop), 'BrodmannArea.Dx']
cellprop1 = cellprop1 %>% 
  rownameToFirstColumn('SampleID') %>%
  dplyr::select(-`P-value`,-Correlation, -RMSE) %>%
  tidyr::gather(Celltype, Fraction, -SampleID, -BrodmannArea.Dx) %>%
  dplyr::arrange(desc(BrodmannArea.Dx))

p = ggplot(cellprop1, aes(x = SampleID, y = Fraction, fill = Celltype, color = BrodmannArea.Dx)) + geom_bar(stat = 'identity')
p = p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
# p = p + facet_grid(.~BrodmannArea.Dx)
# p

p = ggplot(cellprop1, aes(x = Celltype, y = Fraction, color = BrodmannArea.Dx)) + geom_boxplot()
p = p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p
```
### Synapse store
```{r syn.store, cache=TRUE}
CODE = Folder(name = 'MSSM Cell Type Abundance with Reprocessed RNASeq', parentId = parentId)
CODE = synStore(CODE)

cellprop %>%
  rownameToFirstColumn('SampleID') %>%
  write.table(file = 'MSSM_CellFraction_Reproc.tsv', row.names = F, sep = '\t', quote = F)
obj = File('MSSM_CellFraction_Reproc.tsv', name = 'MSSM Cell Fraction', parentId = CODE$properties$id)
obj = synStore(obj, activityName = activityName, activityDescription = activityDescription,
               used = ALL_USED_IDs, executed = thisFile)
```