---
title: "Covariate analysis of MSSM reprocessed counts (with CQN normalisation and covariates adjusted)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knit2synapse) # get the package from devtools::install_github('Sage-Bionetworks/knit2synapse')

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./MSSM_REPROCESSED.Rmd",
                                 parentId = "syn7554308",
                                 entityName = 'MSSM Reprocessed RNASeq (Normalised)')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## It is assumed your working directory is where this file

## Load required libraries
library(CovariateAnalysis) # get the package from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(reshape2)
library(limma)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(RColorBrewer)
library(cqn)

library(synapseClient)
library(knitr)
library(githubr) # get the package from devtools::install_github('brian-bot/githubr')

library(doParallel)
library(foreach)

cl = makeCluster(detectCores()-2)
doParallel::registerDoParallel(cl)

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn7554308'
activityName = 'Covariate analysis';
activityDescription = 'Covariate analysis of MSSM reprocessed RNASeq data';

thisFileName <- 'MSSM_REPROCESSED.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='AMPAD_reprocessing')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Get reprocessed sample counts for temporal cortex
COUNT_ID <- 'syn7994853';
ALL_USED_IDs <- COUNT_ID
COUNT <- read.table(synGet(COUNT_ID)@filePath, header=T, sep='\t', check.names = F)

# Convert rownames of counts from tracking id to ensemble gene id
tmp = data.frame(Gene.ID = rownames(COUNT)) %>%
  dplyr::mutate(ID = Gene.ID) %>%
  tidyr::separate(ID, c('ensembl_gene_id', 'position'), sep = '\\.')
rownames(tmp) = tmp$Gene.ID
rownames(COUNT) = tmp[rownames(COUNT), 'ensembl_gene_id']

# Get sample ids
SampleID = data.frame(SampleID = colnames(COUNT),
                 ID = colnames(COUNT)) %>% 
  tidyr::separate(ID, c('A','B','ID'), sep = '_') %>%
  dplyr::select(SampleID, ID) %>%
  dplyr::mutate(ID = as.numeric(ID))

# Get technical metadata
METADATA_TECH_ID <- 'syn6100548'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_TECH_ID
METADATA_TECH <- read.table(synGet(METADATA_TECH_ID)@filePath,sep=',',header=T, row.names=1) %>%
  group_by(sampleIdentifier) %>%
  top_n(1, -rRNA.rate) %>%
  dplyr::select(sampleIdentifier, BrodmannArea, barcode, individualIdentifier, batch, RIN) %>%
  unique %>%
  ddply(.(barcode), .fun = function(x){
    x = dplyr::mutate(x, batch = paste(batch, collapse = ''), RIN = sum(RIN, na.rm = T))  
  }) 

# Get clinical metadata
METADATA_CLINICAL_ID <- 'syn6101474'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_CLINICAL_ID
METADATA_CLINICAL <- read.csv(synGet(METADATA_CLINICAL_ID)@filePath, header = T)

# Get picard metrics from synapse
METADATA.PICARD_ID <- 'syn8380379'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA.PICARD_ID
METADATA.PICARD_FPATHS <- untar(synGet(METADATA.PICARD_ID)@filePath, list = FALSE, exdir = getwd())
METADATA.PICARD_FPATHS <- untar(synGet(METADATA.PICARD_ID)@filePath, list = TRUE, exdir = getwd())
METADATA.PICARD_FPATHS = METADATA.PICARD_FPATHS[grep('picard.CombinedMetrics.csv', METADATA.PICARD_FPATHS)]
METADATA.PICARD <- plyr::ldply(METADATA.PICARD_FPATHS, read.csv)

METADATA.PICARD$sampleIdentifier = gsub('picard/', '', METADATA.PICARD_FPATHS) %>%
  gsub('.accepted_hits.sort.coord/picard.CombinedMetrics.csv','',.)
colnames(METADATA.PICARD) = gsub('AlignmentSummaryMetrics__','',colnames(METADATA.PICARD))
colnames(METADATA.PICARD) = gsub('RnaSeqMetrics__','',colnames(METADATA.PICARD))

METADATA <- full_join(METADATA_TECH, METADATA_CLINICAL) %>%
  inner_join(METADATA.PICARD)

# Get gene specific parameters from synapse
GENE.PARAM = downloadFile('syn8449369')
ALL_USED_IDs = c(ALL_USED_IDs, 'syn8449369')

GENE.LEN = dplyr::select(GENE.PARAM, ensembl_gene_id, gene.length) %>% unique()
rownames(GENE.LEN) = GENE.LEN$ensembl_gene_id

GENE.GC = dplyr::select(GENE.PARAM, ensembl_gene_id, percentage_gc_content) %>% unique()
rownames(GENE.GC) = GENE.GC$ensembl_gene_id
```

### Data preprocessing
```{r preprocess.data,cache=FALSE, echo=TRUE}
# Fix metadata
METADATA = METADATA %>%
  dplyr::mutate(AOD = gsub('\\+','',AOD),
                BrainRegion = factor(BrodmannArea,
                                     levels = c('BM10', 'BM22', 'BM36', 'BM44'),
                                     labels = c('BM10' = 'FP', 'BM22' = 'STG', 
                                                'BM36' = 'PHG', 'BM44' = 'IFG')),
                Diagnosis = 'NONE', 
                RIN2 = RIN^2)
METADATA$Diagnosis[METADATA$CDR <= 1] = 'ND'
METADATA$Diagnosis[METADATA$CDR == 2 | METADATA$CDR == 3] = 'MD'
METADATA$Diagnosis[METADATA$CDR >= 4] = 'SD'
levels(METADATA$batch)[levels(METADATA$batch) == ''] = 'NoBatch'

METADATA = METADATA %>%
  tidyr::unite(BrainRegion.Diagnosis, BrainRegion, Diagnosis, sep = '.')

METADATA <- METADATA %>%
  filter(!is.na(BrainRegion.Diagnosis), !is.na(PMI), !is.na(RIN))
```

```{r preprocess.data1}
# Match covariates to expression data
indToRetain = intersect(METADATA$sampleIdentifier, colnames(COUNT))
indRemoved = setdiff(colnames(COUNT), METADATA$sampleIdentifier)

COUNT <- COUNT[, indToRetain]

METADATA = as.data.frame(METADATA)
rownames(METADATA) = METADATA$sampleIdentifier
METADATA = METADATA[indToRetain,]
```
`r dim(COUNT)[2]` samples from `r length(unique(METADATA$individualIdentifier))` subjects were obtained from the MSSM cohorts in AMP-AD reprocessed rnaseq project. 

Following `r length(indRemoved)` samples were removed: `r indRemoved`

### Covariate clustering
Determine relationship between covariates
```{r covariates.clustering, cache=TRUE}
primaryVariable <- "BrainRegion.Diagnosis"
FactorCovariates <- c("individualIdentifier", "batch", "RACE", "SEX", "BrainRegion.Diagnosis")
ContCovariates <- c("RIN", "RIN2", "AOD","PMI", "PCT_PF_READS", "PCT_PF_READS_ALIGNED", 
                    "PCT_INTERGENIC_BASES", "PCT_INTRONIC_BASES", 
                    "PCT_RIBOSOMAL_BASES", "PCT_USABLE_BASES")

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- METADATA$sampleIdentifier

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
Correlation/association between covariates at an FDR <= 0.1
```{r covariates.correlation, fig.width=6, fig.height=6}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
draw(COVARIATES.CORRELATION$plot, heatmap_legend_side = 'left')
```

```{r data.explore, fig.width = 12, fig.height = 8}
my.theme = theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1))

# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = RIN)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme

# Age Of Death
p[[2]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = as.numeric(AOD))) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AOD') + my.theme

# PMI
p[[3]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = as.numeric(PMI))) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PMI') + my.theme

# Ribosomal bases
p[[4]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = as.numeric(PCT_RIBOSOMAL_BASES))) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('Fraction of Ribosomal Bases') + my.theme

# Intronic bases
p[[5]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = as.numeric(PCT_INTRONIC_BASES))) + geom_boxplot()
p[[5]] = p[[5]] + ggtitle('Fraction Intronic Bases') + my.theme

# Intergenic bases
p[[6]] = ggplot(COVARIATES, aes(x = BrainRegion.Diagnosis, y = as.numeric(PCT_INTERGENIC_BASES))) + geom_boxplot()
p[[6]] = p[[6]] + ggtitle('Fraction Intergenic Bases') + my.theme

multiplot(plotlist = p, cols = 3)
```
### CPM Normalisation
Preprocess counts matrix using CQN (Conditional Quantile Normlaisation accounting for both GC content and gene length)

Remove genes that have less than 1 cpm counts in at least 50% of samples per BrainRegion x Diagnosis. Also remove genes with missing gene length and percentage GC content
```{r cpmnormalisation}
genesToAnalyze = COVARIATES %>%
  rownameToFirstColumn('SampleID') %>%
  dlply(.(BrainRegion.Diagnosis), .fun = function(mtd, count){
    processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$SampleID],
                                                     MIN_GENE_CPM=1, 
                                                     MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
  processed.counts$filteredExprMatrix$genes
}, COUNT)

genesToAnalyze = unlist(genesToAnalyze) %>% 
  unique() %>%
  intersect(GENE.GC$ensembl_gene_id[!is.na(GENE.GC$percentage_gc_content)]) %>%
  intersect(GENE.LEN$ensembl_gene_id[!is.na(GENE.LEN$gene.length)]) %>%
  setdiff(c("N_unmapped", "N_multimapping", "N_noFeature", "N_ambiguous"))

PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, ], 
                                                 MIN_GENE_CPM=0, 
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples

### Normalisation (with NULL)
Initial normalisation usign cqn and dispersion control with voom (with NULL design)
```{r initial.voom.normalisation, fig.height=5, fig.width=5}
# Compute offset for gene length and gc content
CQN.GENE_EXPRESSION = cqn(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                          x = GENE.GC[rownames(PROCESSED_COUNTS$filteredExprMatrix), 'percentage_gc_content'],
                          lengths = GENE.LEN[rownames(PROCESSED_COUNTS$filteredExprMatrix), 'gene.length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)

# Set gene counts in specific samples that are deviating 3 sd from other samples to 0
log.mat = apply(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  return((x < (mn-3*std.dev)) | (x > (mn+3*std.dev)))
}) %>% t
PROCESSED_COUNTS$filteredExprMatrix$counts[log.mat] = 0

# Re-compute offset for gene length and gc content
CQN.GENE_EXPRESSION = cqn(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                          x = GENE.GC[rownames(PROCESSED_COUNTS$filteredExprMatrix), 'percentage_gc_content'],
                          lengths = GENE.LEN[rownames(PROCESSED_COUNTS$filteredExprMatrix), 'gene.length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)
```
Coexpression of genes 
```{r coexp1, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```
### Outlier analysis
Clustering of voom adjusted data (with NULL design)
```{r decompse.normalise.data1.1, fig.height=5, fig.width=10, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(BrainRegion.Diagnosis, c('Region','Diagnosis'), sep = '\\.') %>% 
  dplyr::mutate(Region = factor(Region), batch = factor(batch))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Diagnosis, size=RIN, shape = SEX))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(.~Region, scales = 'free_y')
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=0)
p
```
Tree based classification of samples
```{r decompse.normalise.data1.2, fig.height=6, fig.width=10, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = (COVARIATES[,c(FactorCovariates), drop = F])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(CQN.GENE_EXPRESSION$y + CQN.GENE_EXPRESSION$offset)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

```{r temp, include = F}
dev.off()
gc()
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates are used to find significant covariates
```{r preadj.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
preAdjustedSigCovars = runPCAandPlotCorrelations(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset, 
                                                 COVARIATES,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```

Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preadj.covariates.plot, fig.width=12, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```
### Normalisation (iterative design)
Since many covariates are correlated, re-normalising and re-adjusting COUNTS with an iterative design matrix.

NOTE:
1. Using a mixed effect model where random effect is chosen as individualIdentifier
2. Adding batch and SEX a priori to variable selection
3. Primary variable of interest BrainRegion.Diagnosis is excluded from the pool of available covariates for selection
```{r iterative.norm, results='asis'}
# Primary variable of interest
postAdjustCovars = c('batch', 'SEX');

# Assign residual covariates
residualCovars = setdiff(preAdjustedSigCovars$significantCovars, 
                         c(postAdjustCovars, primaryVariable, 'individualIdentifier'))
residualSigCovars = preAdjustedSigCovars
covariatesEffects = preAdjustedSigCovars$Effects.significantCovars[residualCovars]
postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects))) %>% unique()

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 20){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects and individualIdentifier as random effect'))

  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]

  # Estimate voom weights for dispersion control
  VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix$counts, design=DM1$design, plot=F)

  # Calculate correlation values of random effects
  VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset
  correlation = parallelDuplicateCorrelation(VOOM.GENE_EXPRESSION,
                                             block = COVARIATES$individualIdentifier)
  
  # Re-estimate voom weights
  VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                              design=DM1$design, plot=F, 
                              block = COVARIATES$individualIdentifier, 
                              correlation = correlation$cor)

  # Fit linear model using new weights and new design
  VOOM.ADJUSTED.FIT = lmFit(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset,
                            design = DM1$design,
                            weights = VOOM.GENE_EXPRESSION$weights,
                            block = COVARIATES$individualIdentifier, 
                            correlation = correlation$cor)

  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,
                                                CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset)

  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates),
                         c(postAdjustCovars, primaryVariable, 'individualIdentifier'))

  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                COVARIATES[, residCovars, drop=F], 
                                                'adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)

  # Add postadjusted covariates (if any)
  residCovars = setdiff(residualSigCovars$significantCovars, 
                        c(postAdjustCovars, primaryVariable, 'individualIdentifier'))
  covariatesEffects = residualSigCovars$Effects.significantCovars[residCovars]

  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}

modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```

### Sanity check
```{r residual.adj, fig.width=12, fig.height=8}
# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                              COVARIATES,
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)

residualSigCovars[["PC_res"]][[2]]$plotData
```
Coexpression of genes 
```{r coexp2, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(RESIDUAL.GENE_EXPRESSION))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Clustering residual data
```{r decompse.normalise.data2.1, fig.height=10, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION, scale.=T, center = T)

# Plot first 4 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(BrainRegion.Diagnosis, c('Region','Diagnosis'), sep = '\\.') %>% 
  dplyr::mutate(Region = factor(Region))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Diagnosis, size=RIN, shape = SEX))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(.~Region, scales = 'free_y')
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=0)
p
```

```{r decompse.normalise.data2.2, fig.height=8, fig.width=12, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,FactorCovariates])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(RESIDUAL.GENE_EXPRESSION)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

```{r temp1, include=F}
dev.off()
gc()
```

### Differential expression analysis (with BrainRegion.Diagnosis as primary variable)
Differential expression is performed on the primary variable by controlling for covariates identified above

Interpretation of BrainRegion.Diagnosis
1. FP.MD: Mild dementia samples from frontal pole with CDR equal to 2 or 3
2. FP.ND: No dementia samples from frontal pole with CDR less than 2
3. FP.SD: Severe dementia samples from frontal pole with CDR grater than 3
4. IFG.MD: Mild dementia samples from inferior frontal gyrus with CDR equal to 2 or 3
5. IFG.ND: No dementia samples from inferior frontal gyrus with CDR less than 2
6. IFG.SD: Severe dementia samples from inferior frontal gyrus with CDR grater than 3
7. PHG.MD: Mild dementia samples from para hippocampal gyrus with CDR equal to 2 or 3
8. PHG.ND: No dementia samples from para hippocampal gyrus with CDR less than 2
9. PHG.SD: Severe dementia samples from para hippocampal gyrus with CDR grater than 3
10. STG.MD: Mild dementia samples from superior temporal gyrus with CDR equal to 2 or 3
11. STG.ND: No dementia samples from superior temporal gyrus with CDR less than 2
12. STG.SD: Severe dementia samples from superior temporal gyrus with CDR grater than 3

Genes that are differentially expressed at an FDR <= 0.05 are
```{r diff.exp, fig.height=15, fig.width=13}
# Get design matrix
DESIGN = getDesignMatrix(COVARIATES[, c(primaryVariable, postAdjustCovars), drop = F], Intercept = F)
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]

# Estimate voom weights for dispersion control
VOOM.WEIGHTS = voom(PROCESSED_COUNTS$filteredExprMatrix$counts, design=DESIGN$design, plot=F)

# Calculate correlation values of random effects
VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset
correlation = parallelDuplicateCorrelation(VOOM.WEIGHTS, block = COVARIATES$individualIdentifier)
  
# Re-estimate voom weights
VOOM.WEIGHTS = voom(PROCESSED_COUNTS$filteredExprMatrix$counts,
                    design=DESIGN$design, plot=F,
                    block = COVARIATES$individualIdentifier, 
                    correlation = correlation$cor)

# Fit linear model using new weights and new design
VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset
FIT = lmFit(VOOM.WEIGHTS)

# Fit contrast
cntr = c()
for (region in c('FP', 'STG', 'PHG', 'IFG')){
  tmp = combn(paste0('BrainRegion.Diagnosis', region, '.', c('ND','MD','SD')),2)
  cntr = c(cntr, apply(tmp, 2, paste, collapse = '-'))
}

contrast = makeContrasts(contrasts=cntr,
                         levels = colnames(FIT$coefficients))
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)

# Get differnetial expression
DE = lapply(1:12, function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = T) %>%
    rownameToFirstColumn('ensembl_gene_id')
}, FIT.CONTR)
names(DE) = colnames(contrast)

DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('BrainRegion.Diagnosis','',Comparison),
                Study = 'MSSM',
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  tidyr::separate(Comparison, c('from.state','to.state'), sep = '-') %>%
  tidyr::separate(from.state, c('Region','reference'), sep = '\\.') %>%
  tidyr::separate(to.state, c('Region','against'), sep = '\\.') %>%
  tidyr::unite(Comparison, reference, against, sep = '-') %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gc_content, gene.length) %>%
              unique())
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, Comparison, Direction, Region) %>%
  group_by(Region, Comparison, Direction) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05) 

tmp1 = DE %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::select(ensembl_gene_id, Comparison, Direction, Region) %>%
  group_by(Region, Comparison, Direction) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id))) %>%
  spread(Direction, FDR_0_05_FC_1.2) 

kable(full_join(tmp %>%
                  dplyr::rename(DOWN_FDR_0_05 = DOWN, UP_FDR_0_05 = UP),
                tmp1 %>%
                  dplyr::rename(DOWN_FDR_0_05_FC_1.2 = DOWN, UP_FDR_0_05_FC_1.2 = UP)))

p = ggplot(DE, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red'))
p = p + facet_grid(Region+.~Comparison, scales = 'fixed')
p
```

### Associate differential expression results with gc content, gene length and average expression
```{r associate.de, fig.height=15, fig.width=15}
pl = list()
pl[[1]] = ggplot(DE %>% dplyr::filter(Comparison == 'ND-SD'), 
                 aes(x = log10(gene.length), y = logFC, color = Direction)) + geom_point() 
pl[[1]] = pl[[1]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = log10(gene.length), y = logFC))
pl[[1]] = pl[[1]] + facet_grid(Region~.+Comparison) + scale_color_manual(values = c('green','grey','red'))
pl[[1]] = pl[[1]] + theme(legend.position = 'top')

pl[[2]] = ggplot(DE %>% dplyr::filter(Comparison == 'ND-SD'), 
                 aes(x = percentage_gc_content, y = logFC, color = Direction)) + geom_point()
pl[[2]] = pl[[2]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = percentage_gc_content, y = logFC))
pl[[2]] = pl[[2]] + facet_grid(Region~.+Comparison) + scale_color_manual(values = c('green','grey','red'))
pl[[2]] = pl[[2]] + theme(legend.position = 'top')

pl[[3]] = ggplot(DE %>% dplyr::filter(Comparison == 'ND-SD'), 
                 aes(x = AveExpr, y = logFC, color = Direction)) + geom_point()
pl[[3]] = pl[[3]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = AveExpr, y = logFC))
pl[[3]] = pl[[3]] + facet_grid(Region~.+Comparison) + scale_color_manual(values = c('green','grey','red'))
pl[[3]] = pl[[3]] + theme(legend.position = 'top')

multiplot(plotlist = pl, cols = 3)
```

### SVA Adjustments for eQTL analysis
Conditioned on primary variable (BrainRegion.Diagnosis) and identified covariates estimate surrogate variables using SVA package
```{r sva.adjust}
# Get (null) design matrix
MODEL0 = DESIGN$design[,-grep(primaryVariable,colnames(DESIGN$design))] # Get null model by removing variable of interest
MODEL0 = MODEL0[,linColumnFinder(MODEL0)$indepCols]
MODEL1 = DESIGN$design
MODEL1 = MODEL1[,linColumnFinder(MODEL1)$indepCols]

# Calculate the number of surrogate variables
NUM.SV = sva::num.sv(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset, 
                     MODEL1, method = 'be', seed = 123456, B = 30)

# Estimate surrogate variables
SURR.VAR = sva::sva(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset,
                    MODEL1, MODEL0, n.sv = NUM.SV, B = 30)$sv

colnames(SURR.VAR) = paste0('SV',1:dim(SURR.VAR)[2])
rownames(SURR.VAR) = rownames(MODEL1)

# Estimate voom weights for dispersion control
VOOM.SVA.WEIGHTS = voom(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                        design=cbind(MODEL1, SURR.VAR), plot=F)

# Calculate correlation values of random effects
VOOM.SVA.WEIGHTS$E = CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset
correlation = parallelDuplicateCorrelation(VOOM.SVA.WEIGHTS, block = COVARIATES$individualIdentifier)

# Re-estimate voom weights
VOOM.SVA.WEIGHTS = voom(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                        design = cbind(MODEL1, SURR.VAR), plot=F,
                        block = COVARIATES$individualIdentifier, 
                        correlation = correlation$cor)

# Fit linear model using new weights and new design
VOOM.SVA.WEIGHTS$E = CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset
VOOM.SVA.FIT = lmFit(VOOM.SVA.WEIGHTS)

# Residuals after normalisation
RESIDUAL.SVA.GENE_EXPRESSION = residuals.MArrayLM(VOOM.SVA.FIT,
                                                  CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset)
```

### Covariates Adjustments for network analysis
Identified covariates are regressed out from the expression matrix for network analysis
```{r network.adjust}
# Get design matrix
DESIGN.NET = getDesignMatrix(COVARIATES[, postAdjustCovars, drop = F], Intercept = F)
DESIGN.NET = DESIGN.NET$design[,linColumnFinder(DESIGN.NET$design)$indepCols]

# Estimate voom weights for dispersion control
VOOM.NET.WEIGHTS = voom(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                        design=DESIGN.NET, plot=F)

# Calculate correlation values of random effects
VOOM.NET.WEIGHTS$E = CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset
correlation = parallelDuplicateCorrelation(VOOM.NET.WEIGHTS, block = COVARIATES$individualIdentifier)

# Re-estimate voom weights
VOOM.NET.WEIGHTS = voom(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                        design = DESIGN.NET, plot=F,
                        block = COVARIATES$individualIdentifier, 
                        correlation = correlation$cor)

# Fit linear model using new weights and new design
VOOM.NET.WEIGHTS$E = CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset
VOOM.NET.FIT = lmFit(VOOM.NET.WEIGHTS)

# Residuals after normalisation
RESIDUAL.NET.GENE_EXPRESSION = residuals.MArrayLM(VOOM.NET.FIT,
                                                  CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset)
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Set annotations
all.annotations = list(
  dataType = 'mRNA',
  dataSubType = 'geneExp',
  summaryLevel = 'gene',
  assay	 = 'Reprocessed RNAseq',
  
  tissueTypeAbrv	= 'FP, STG, PHG, IFG', 
  study = 'MSSMRNASeq', 
  center = 'MSSM',
  
  organism = 'HomoSapiens',
  consortium	= 'AMP-AD',
   
  normalizationStatus	= TRUE,
  modelSystem	= FALSE,
  
  normalizationType	= 'CQN'
)

# Code
CODE <- Folder(name = "MSSM Reprocessed RNASeq (Normalised)", parentId = parentId)
annotations(CODE) = all.annotations
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'SampleID')
write.table(COVARIATES, file = 'MSSM_FP_STG_PHG_IFG_Covariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = File('MSSM_FP_STG_PHG_IFG_Covariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
annotations(COV_OBJ) = annotations(CODE)
annotations(COV_OBJ)$dataSubType = 'covariates'
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                      executed = thisFile, activityDescription = activityDescription)

# Store filtered counts
PROCESSED_COUNTS$filteredExprMatrix$counts %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'MSSM_FP_STG_PHG_IFG_Counts.tsv', sep = '\t', row.names=F, quote=F)
COUNT_OBJ = File('MSSM_FP_STG_PHG_IFG_Counts.tsv', name = 'Counts (filtered raw)', parentId = CODE$properties$id)
annotations(COUNT_OBJ) = annotations(CODE)
annotations(COUNT_OBJ)$dataSubType = 'filteredCounts'
COUNT_OBJ = synStore(COUNT_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store logCPM
CQN.GENE_EXPRESSION$y %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'MSSM_FP_STG_PHG_IFG_logCPM.tsv', sep = '\t', row.names=F, quote=F)
LCOUNT_OBJ = File('MSSM_FP_STG_PHG_IFG_logCPM.tsv', name = 'Counts (filtered logCPM)', parentId = CODE$properties$id)
annotations(LCOUNT_OBJ) = annotations(CODE)
annotations(LCOUNT_OBJ)$dataSubType = 'filteredLCPM'
LCOUNT_OBJ = synStore(LCOUNT_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store cqn offsets
CQN.GENE_EXPRESSION$offset %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'MSSM_FP_STG_PHG_IFG_offset.tsv', sep = '\t', row.names=F, quote=F)
OFFSET_OBJ = File('MSSM_FP_STG_PHG_IFG_offset.tsv', name = 'Gene length and GC content offset', parentId = CODE$properties$id)
annotations(OFFSET_OBJ) = annotations(CODE)
annotations(OFFSET_OBJ)$dataSubType = 'offset'
OFFSET_OBJ = synStore(OFFSET_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store design matrix
MODEL1 %>%
  rownameToFirstColumn('SampleID') %>%
  write.table(file = 'MSSM_FP_STG_PHG_IFG_Design.tsv', sep = '\t', row.names=F, quote=F)
DM_OBJ = File('MSSM_FP_STG_PHG_IFG_Design.tsv', name = 'Design Matrix', parentId = CODE$properties$id)
annotations(DM_OBJ) = annotations(CODE)
annotations(DM_OBJ)$dataSubType = 'designMatrix'
DM_OBJ = synStore(DM_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store surrogate variables
SURR.VAR %>%
  rownameToFirstColumn('SampleID') %>%
  write.table(file = 'MSSM_FP_STG_PHG_IFG_SurVar.tsv', sep = '\t', row.names=F, quote=F)
SURR_OBJ = File('MSSM_FP_STG_PHG_IFG_SurVar.tsv', name = 'Surrogate Variables', parentId = CODE$properties$id)
annotations(SURR_OBJ) = annotations(CODE)
annotations(SURR_OBJ)$dataSubType = 'surrogateVariables'
SURR_OBJ = synStore(SURR_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store residual gene expression for eQTL analysis
RESIDUAL.SVA.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'MSSM_FP_STG_PHG_IFG_eQTLResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
eEXP_OBJ = File('MSSM_FP_STG_PHG_IFG_eQTLResidualExpression.tsv', 
                name = 'Normalised, covariates and surrogate variable adjusted residual expression (for eQTL)',
                parentId = CODE$properties$id)
annotations(eEXP_OBJ) = annotations(CODE)
annotations(eEXP_OBJ)$dataSubType = 'residualGeneExpForeQTL'
eEXP_OBJ = synStore(eEXP_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store residual gene expression for network analysis
RESIDUAL.NET.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'MSSM_FP_STG_PHG_IFG_netResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
nEXP_OBJ = File('MSSM_FP_STG_PHG_IFG_netResidualExpression.tsv', 
                name = 'Normalised, covariates removed residual expression (for network analysis)', 
                parentId = CODE$properties$id)
annotations(nEXP_OBJ) = annotations(CODE)
annotations(nEXP_OBJ)$dataSubType = 'residualGeneExpForNetAnlz'
nEXP_OBJ = synStore(nEXP_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store differential expression results
write.table(DE, file = 'MSSM_FP_STG_PHG_IFG_DiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('MSSM_FP_STG_PHG_IFG_DiffExpression.tsv', 
                name = 'Differential Expression Results (BrainRegion.Diagnosis)', 
                parentId = CODE$properties$id)
annotations(DEXP_OBJ) = annotations(CODE)
annotations(DEXP_OBJ)$dataSubType = 'diffExp'
DEXP_OBJ = synStore(DEXP_OBJ, activity = synGetActivity(COV_OBJ$properties$id))
stopCluster(cl)
```
|  *Results*                                  |  *SynapseID*                     |
|  -----------------------------------------  |   ---------                      |
|  Covariates                                 |  `r COV_OBJ$properties$id`       |
|  Counts (raw)                               |  `r COUNT_OBJ$properties$id`     |
|  Counts (lcpm)                              |  `r LCOUNT_OBJ$properties$id`    |
|  Offset (for gene length and gc content)    |  `r OFFSET_OBJ$properties$id`    |
|  Design Matrix                              |  `r DM_OBJ$properties$id`        |
|  Surrogate Variables                        |  `r SURR_OBJ$properties$id`      |
|  Residual Expression (for eQTL)             |  `r eEXP_OBJ$properties$id`      |
|  Residual Expression (for network analysis) |  `r nEXP_OBJ$properties$id`      |
|  Differential Expression                    |  `r DEXP_OBJ$properties$id`      |
