---
title: "Covariates analysis of MSBB RNASeq data"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(knit2synapse)
library(synapseClient)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./MSSM_RNASeq.Rmd",
                                 parentId ="syn5570248",
                                 entityName = 'MSBB RNASeq')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file

### Clear R console screen output
cat("\014") 

## Load required libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(reshape2)
library(limma)
library(RColorBrewer)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(xlsx)

# library(doParallel)
# library(foreach)
# 
# cl = makeCluster(12)
# registerDoParallel(cl)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# Source required R files
source('../R/lib/parallelDuplicateCorrelation.R')

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn5570248';
activityName = 'Covariate analysis';
activityDescription = 'Covariate analysis of MSBB RNASeq samples';

thisFileName <- 'MSSM_RNASeq.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='AMPAD')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
Data were obtained from AMP-AD portal of MSSM, and following brain regions were analysed for differential expression between case and control:
  | Region | Name                      | 
  | :----- | ------------------------- | 
  | BM_10  | Frontal Pole (FP)         |
  | BM_22  | Superior Temporal Gyrus   |
  | BM_36  | Parahipocampal Gyrus (PHG)|
  | BM_44  | inferior frontal Gyrus (IFG)|
  
* BM 10: frontopolar prefrontal cortex, rostrolateral prefrontal cortex, or anterior prefrontal cortex, 
* BM 22: In humans it corresponds approximately to the lateral and caudal two thirds of the superior temporal gyrus, 
* BM 36: Ectorhinal area 36 is a subdivision of the cytoarchitecturally defined temporal region of cerebral cortex,
* BM 44: pars opercularis (of the inferior frontal gyrus), and it refers to a subdivision of the frontal region of cerebral cortex.

Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download expression data
COUNT_ID <- 'syn5898488'
ALL_USED_IDs <- COUNT_ID
COUNT_OBJ <- synGet(COUNT_ID)
COUNT <- fread(COUNT_OBJ@filePath, data.table = F)
rownames(COUNT) = COUNT$Geneid

# Get technical metadata
METADATA_ID <- 'syn5898489'
METADATA_OBJ <- synGet(METADATA_ID)
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_ID
METADATA <- read.table(METADATA_OBJ@filePath,sep='\t',header=T, row.names=1)

# Get clinical metadata
METADATA_TECHNICAL_ID <- 'syn5475828'
METADATA_TECHNICAL_OBJ <- synGet(METADATA_TECHNICAL_ID)
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_TECHNICAL_ID
METADATA_TECHNICAL <- read.xlsx(METADATA_TECHNICAL_OBJ@filePath, sheetIndex = 1)

# Manually fix 4 sample IDs
METADATA_TECHNICAL[which(METADATA_TECHNICAL$BB == 1132), "BM36.Fnl"] = 397
METADATA_TECHNICAL[which(METADATA_TECHNICAL$BB == 1117), "BM22.Fnl"] = 144
METADATA_TECHNICAL[which(METADATA_TECHNICAL$BB == 1320), "BM22.Fnl"] = 159
METADATA_TECHNICAL[which(METADATA_TECHNICAL$BB == 1475), "BM22.Fnl"] = 186

METADATA = left_join(METADATA %>%
                       dplyr::mutate(ID = gsub('hB_RNA_','',NewBarcode),
                                     ID = as.numeric(ID)),
                     METADATA_TECHNICAL %>%
                       dplyr::select(BB, BM.44.Barcode, BM36.Fnl, BM22.Fnl, BM10.Fnl, PMI, RACE, 
                                     AOD, CDR, SEX, NP.1, PlaqueMean, bbscore) %>%
                       dplyr::rename(BM_44 = BM.44.Barcode,
                                     BM_10 = BM10.Fnl,
                                     BM_22 = BM22.Fnl,
                                     BM_36 = BM36.Fnl) %>%
                       tidyr::gather(BrainRegion, ID, -BB, -PMI, -RACE, 
                                     -AOD, -CDR, -SEX, -NP.1, -PlaqueMean, -bbscore) %>%
                       filter(!is.na(ID)) %>%
                       dplyr::mutate(ID = as.numeric(ID)))
rownames(METADATA) = METADATA$NewBarcode

# Filter data
ind <- intersect(rownames(METADATA), colnames(COUNT))
ind <- setdiff(ind, 'hB_RNA_10622')
METADATA <- METADATA[ind,]
COUNT <- COUNT[, ind]
```
### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering, cache=TRUE}
# Max out age at 90
METADATA$AOD = gsub('\\+','',as.character(METADATA$AOD))

# Create RIN^2 column
METADATA$RIN2 = METADATA$RIN^2

# Create diagnosis column (classified into 3 groups ND, MD, SD)
METADATA$Dx = 'ND'
METADATA$Dx[METADATA$CDR == 1 | METADATA$CDR == 2] = 'MD'
METADATA$Dx[METADATA$CDR >2 ] = 'SD'

# Create brain region x diagnosis column
METADATA$BrainRegion.Dx = paste(METADATA$BrainRegion, METADATA$Dx, sep = '.')

# Select covariates of interest
FactorCovariates <- c('RACE', 'SEX', 'BrainRegion.Dx', 'Batch', 'BB')
ContCovariates <- c('PMI', 'AOD', 'RIN', 'RIN2')

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- METADATA$NewBarcode

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)

# Arrange expression and covariates
COUNT = COUNT[, rownames(COVARIATES)]
```
Covariate correlation
```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```

```{r dataExplore, fig.width = 10, fig.height = 10}
# RIN
p = list()
p[[1]] = ggplot(METADATA, aes(x = Dx, y = RIN, fill = BrainRegion)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + theme(legend.position = 'top')

# AgeAtDeath
p[[2]] = ggplot(METADATA, aes(x = Dx, y = as.numeric(AOD), fill = BrainRegion)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AgeOfDeath') + theme(legend.position = 'top')

# Mean Plaque
p[[3]] = ggplot(METADATA, aes(x = Dx, y = as.numeric(PlaqueMean), fill = BrainRegion)) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PlaqueMean') + theme(legend.position = 'top')

# Mean Plaque
p[[4]] = ggplot(METADATA, aes(x = Dx, y = as.numeric(NP.1), fill = BrainRegion)) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('NP.1') + theme(legend.position = 'top')

multiplot(plotlist = p, cols = 2)

# Sex
vcd::mosaic(~ SEX + Dx + BrainRegion, data = METADATA)
```


### CPM Normalisation
Preprocess counts matrix and metadata. Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of samples.
```{r cpmnormalisation, cache=TRUE}
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=1, MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
```
### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)
```{r initial.voom.normalisation}
# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)
```
### Outlier analysis
Clustering of initial normalised data (with NULL design)
```{r decompse.normalise.data, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E, scale.=T)

# Plot first 2 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'ID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=BrainRegion.Dx, shape=SEX, size=PMI))
p <- p + theme_bw() + theme(legend.position="top") + scale_shape_manual(values = 1:7)
# p <- p + geom_text(aes(label= ID), size=4, hjust=1)
p

COVARIATES.tmp = data.matrix(COVARIATES)
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(VOOM.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);
tmp = gsub('\\..*','',cols)
WGCNA::plotDendroAndColors(tree, 
                           colors = tmp, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
dev.off()
```
### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
COVARIATES.tmp = COVARIATES
COVARIATES.tmp[, FactorCovariates] = sapply(COVARIATES.tmp[, FactorCovariates], as.numeric)
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, COVARIATES.tmp,'NULL design(voom-normalized)', isKeyPlot=TRUE, MIN_PVE_PCT_PC = 1)
```
Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative design)
Since many covariates are correlated, re-normalising COVARIATES with an iterative design matrix. Here BB (Donor Information) is chosen as random effect
```{r iterativeNorm}
# Primary variable of interest
postAdjustCovars = 'BrainRegion.Dx'

# Assign residual covariates
residualSigCovars = preAdjustedSigCovars

covariatesEffects = residualSigCovars$Effects.significantCovars[setdiff(residualSigCovars$significantCovars, c(postAdjustCovars, 'BB'))]

postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 100){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
  
  # Estimate correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F)
  correlation <- parallelDuplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION, block=COVARIATES$BB)
  
  # Re-calculate voom weights with correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F,
                                       block=COVARIATES$BB, correlation = correlation$cor)

  # Re-calculate linear fit using new weights and new design
  VOOM.ADJUSTED.FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), c(postAdjustCovars, 'BB'))
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                COVARIATES.tmp[, residCovars, drop=F], 'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  covariatesEffects = residualSigCovars$Effects.significantCovars[setdiff(residualSigCovars$significantCovars, c(postAdjustCovars, 'BB'))]

  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`
### Sanity check
```{r residualSigCovars.manual, fig.width=20, fig.height=12}
residualSigCovars[["PC_res"]][[2]]$plotData
```
### Residual calculation
Calculate weighted residuals and add back "BrainRegion.Dx" to the residuals
```{r varsToAddBack}
# Add variable of interest back to the residuals
varsToAddIn = grep("BrainRegion.Dx", colnames(DM1$design), value = T)
RESIDUAL.GENE_EXPRESSION = RESIDUAL.GENE_EXPRESSION + 
  VOOM.ADJUSTED.FIT$coefficients[,varsToAddIn] %*% t(DM1$design[,varsToAddIn])
```

### Clustering residual data
```{r decompse.normalise.data2, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION, scale.=T)

# Plot first 2 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'ID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=factor(BrainRegion.Dx),size=as.numeric(PMI)))
p <- p + theme_bw() + theme(legend.position="top") 
# p <- p + geom_text(aes(label= ID), size=4, hjust=0)
p

COVARIATES.tmp = data.matrix(COVARIATES)
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(RESIDUAL.GENE_EXPRESSION)))
cols = WGCNA::labels2colors(COVARIATES.tmp);
tmp = gsub('\\..*','',cols)
WGCNA::plotDendroAndColors(tree, 
                           colors = tmp, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```
```{r get.gene.symbols, cache=TRUE, eval=FALSE}
# Define biomart object
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", host = )

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", values = rownames(RESIDUAL.GENE_EXPRESSION),
                       mart = mart)
RESIDUAL.GENE_EXPRESSION <- left_join(rownameToFirstColumn(RESIDUAL.GENE_EXPRESSION, "ensembl_gene_id"),
                                      Ensemble2HGNC) %>%
  dplyr::select(ensembl_gene_id, hgnc_symbol, one_of(rownames(COVARIATES)))
```
```{r summary, results='asis', echo=FALSE}
writeLines('Number of samples in each category are')
tmp = as.data.frame(summary(COVARIATES$BrainRegion.Dx))
colnames(tmp) = 'No. of samples'
kable(tmp)
```
### Differential expression analysis
Genes that are differentially expressed at an FDR <= 0.05 are
```{r diffExp, fig.height=10, fig.width=18}
# Fit contrast
contrast = makeContrasts(contrasts=c("BrainRegion.DxBM_10.ND-BrainRegion.DxBM_10.MD", 
                                     "BrainRegion.DxBM_10.ND-BrainRegion.DxBM_10.SD", 
                                     "BrainRegion.DxBM_10.MD-BrainRegion.DxBM_10.SD",
                                     "BrainRegion.DxBM_22.ND-BrainRegion.DxBM_22.MD", 
                                     "BrainRegion.DxBM_22.ND-BrainRegion.DxBM_22.SD", 
                                     "BrainRegion.DxBM_22.MD-BrainRegion.DxBM_22.SD",
                                     "BrainRegion.DxBM_36.ND-BrainRegion.DxBM_36.MD", 
                                     "BrainRegion.DxBM_36.ND-BrainRegion.DxBM_36.SD", 
                                     "BrainRegion.DxBM_36.MD-BrainRegion.DxBM_36.SD",
                                     "BrainRegion.DxBM_44.ND-BrainRegion.DxBM_44.MD", 
                                     "BrainRegion.DxBM_44.ND-BrainRegion.DxBM_44.SD", 
                                     "BrainRegion.DxBM_44.MD-BrainRegion.DxBM_44.SD"), 
                         levels = colnames(VOOM.ADJUSTED.FIT$coefficients))
FIT = contrasts.fit(VOOM.ADJUSTED.FIT, contrasts=contrast)
FIT = eBayes(FIT)

# Get differnetial expression
DE = list()
p = list()
for (i in 1:dim(contrast)[2]){
  DE[[i]] = topTable(FIT, coef=i, number = dim(VOOM.ADJUSTED.GENE_EXPRESSION$E)[1])
  DE[[i]] = rownameToFirstColumn(DE[[i]], "ensembl_gene_id")
  p[[i]] <- ggplot(DE[[i]], aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()
  p[[i]] <- p[[i]] + ggtitle(gsub('BrainRegion.Dx','',colnames(contrast)[i])) + xlim(c(-1.5,1.5)) + ylim(c(0,6))
  p[[i]] <- p[[i]] + geom_hline(yintercept = -log10(0.05), color = 'red')
}
multiplot(plotlist=p, cols = 3)
names(DE) = gsub('BrainRegion.Dx','',colnames(contrast))

DE = DE %>% ldply(.id = 'Comparison')

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(Comparison, ensembl_gene_id) %>%
  group_by(Comparison) %>%
  dplyr::summarise(nDiffExp = length(unique(ensembl_gene_id)), 
            DiffExpGenes = paste(unique(ensembl_gene_id), collapse = ','))

kable(tmp)
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Code
CODE <- Folder(name = "MSBB RNASeq", parentId = parentId)
CODE <- synStore(CODE)

# Expression
expr = rownameToFirstColumn(RESIDUAL.GENE_EXPRESSION, 'EnsembleIDs')
write.table(expr, file='expression.tsv', sep = '\t', row.names=F, quote=F)
obj = File('expression.tsv', 
           name = 'Voom Adjusted Weighted Residual Expression (BrainRegion.Dx added)', 
           parentId = CODE$properties$id)
expr_obj = synStore(obj,
               used = ALL_USED_IDs, 
               activityName = activityName, 
               executed = thisFile, 
               activityDescription = activityDescription)
  
# Covariates
covar = rownameToFirstColumn(COVARIATES, 'SampleId')
write.table(covar, file='covariates.tsv', sep = '\t', row.names=F, quote=F)
obj = File('covariates.tsv', 
           name = 'Covariates', 
           parentId = CODE$properties$id)
covar_obj = synStore(obj,
               used = ALL_USED_IDs, 
               activityName = activityName, 
               executed = thisFile, 
               activityDescription = activityDescription)

# Store differential expression results
write.table(DE, file = 'diffExpression.tsv', sep = '\t', row.names=F, quote=F)
dexp_obj = File('diffExpression.tsv', name = 'Differential Expression Analysis', parentId = CODE$properties$id)
dexp_obj = synStore(dexp_obj, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)
```
|  *Results*            |  *SynapseID*                    |
|  -------              |   ---------                     |
|  Covariates           |  `r covar_obj$properties$id`    |
|  Residual Expression  |  `r expr_obj$properties$id`     |
|  Differential Expression  |  `r dexp_obj$properties$id`      |