---
title: "Covariates analysis of MSBB RNASeq data of region BM 36 (realigned from source FASTQ with STAR-HTSeq)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./MSSM_BM_36_RNASeq_Realigned_StarHTSeq.Rmd",
                                 parentId ="syn5570248",
                                 entityName = 'MSBB BM 36 RNASeq RealignedFromSource StarHTSeq')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file

### Clear R console screen output
cat("\014") 

## Load required libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(reshape2)
library(limma)
library(RColorBrewer)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(xlsx)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn5570248';
activityName = 'Covariate analysis';
activityDescription = 'Covariate analysis of MSBB BM 36 RNASeq samples';

thisFileName <- 'MSSM_BM_36_RNASeq_Realigned_StarHTSeq.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='AMPAD')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```

Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download expression data
COUNT_ID <- 'syn7224673'
ALL_USED_IDs <- COUNT_ID
COUNT <- read.table(synGet(COUNT_ID)@filePath, quote = '', header = T, sep = '\t')

# Get sample ids
SampleID = data.frame(SampleID = colnames(COUNT),
                      ID = colnames(COUNT)) %>% 
  tidyr::separate(ID, c('A','B','ID', 'C'), sep = '_') 

SampleID = SampleID %>%
  dplyr::select(SampleID, ID) %>%
  dplyr::mutate(ID = as.numeric(ID),
                Sample.ID = SampleID,
                SampleID = paste0('BM_36_',ID))
  
# Sum counts from multiple lanes
COUNT = dlply(SampleID, .(ID), .fun = function(x, count){
  rowSums(count[,x$Sample.ID,drop=F]) %>%
    rownameToFirstColumn('Gene.ID') %>%
    plyr::rename(c('DF' = paste0('BM_36_',unique(x$ID)))) %>%
    dplyr::select(-Gene.ID)
}, COUNT) %>%
  Reduce(cbind, .)
COUNT = data.matrix(COUNT)

# # Get sample ids that are to be retained (data from Noam on 29th June) 
# tmp = read.table('MSSM_SampleIDs.txt', header = T)
# removeID = setdiff(SampleID$ID, tmp$x)
# writeLines(paste('Removed',length(removeID),'samples from MSSM'))
# SampleID = filter(SampleID, ID %in% tmp$x)

# Get technical metadata
METADATA_TECHNICAL_ID <- 'syn5898489'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_TECHNICAL_ID
METADATA_TECHNICAL <- read.table(synGet(METADATA_ID)@filePath,sep='\t',header=T, row.names=1) %>%
  rownameToFirstColumn('OldBarcode')

# Get clinical metadata
METADATA_CLINICAL_ID <- 'syn5475828'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_CLINICAL_ID
METADATA_CLINICAL <- read.xlsx(synGet(METADATA_CLINICAL_ID)@filePath, sheetIndex = 1)

# Manually fix 4 sample IDs
METADATA_CLINICAL[which(METADATA_CLINICAL$BB == 1132), "BM36.Fnl"] = 397
METADATA_CLINICAL[which(METADATA_CLINICAL$BB == 1117), "BM22.Fnl"] = 144
METADATA_CLINICAL[which(METADATA_CLINICAL$BB == 1320), "BM22.Fnl"] = 159
METADATA_CLINICAL[which(METADATA_CLINICAL$BB == 1475), "BM22.Fnl"] = 186

METADATA = left_join(METADATA_TECHNICAL %>%
                       dplyr::mutate(ID = gsub('hB_RNA_','',NewBarcode),
                                     ID = as.numeric(ID)),
                     METADATA_CLINICAL %>%
                       dplyr::select(BB, BM36.Fnl, PMI, RACE, 
                                     AOD, CDR, SEX, NP.1, PlaqueMean, bbscore) %>%
                       dplyr::rename(BM_36 =  BM36.Fnl) %>%
                       tidyr::gather(BrainRegion, ID, -BB, -PMI, -RACE, 
                                     -AOD, -CDR, -SEX, -NP.1, -PlaqueMean, -bbscore) %>%
                       filter(!is.na(ID)) %>%
                       dplyr::mutate(ID = as.numeric(ID))) %>%
  inner_join(SampleID %>% dplyr::select(-Sample.ID)) %>%
  filter(!is.na(SampleID)) %>%
  filter(!(ID %in% c(335, 338, 339, 341))) %>%
  unique
rownames(METADATA) = METADATA$SampleID

# Filter data
ind <- intersect(METADATA$SampleID, colnames(COUNT))
indRemoved <- setdiff(colnames(COUNT), METADATA$SampleID)
# ind <- setdiff(ind, 'hB_RNA_10622')
METADATA <- METADATA[ind,]
COUNT <- COUNT[, ind]
```

### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering, cache=TRUE}
METADATA$AOD = gsub('\\+','',as.character(METADATA$AOD))
METADATA$RIN2 = METADATA$RIN^2

FactorCovariates <- c('RACE','SEX','CDR','Batch')
ContCovariates <- c('PMI','AOD','PlaqueMean','RIN', 'RIN2')

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- rownames(METADATA)

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)

rownames(COVARIATES) <- rownames(METADATA)
```
Covariate correlation
```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```
### CPM Normalisation
Preprocess counts matrix and metadata. Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of samples.
```{r cpmnormalisation, cache=TRUE}
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=1, MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
```
### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)
```{r initial.voom.normalisation}
# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)
```
Coexpression of genes
```{r coexp1}
cr = cor(t(VOOM.GENE_EXPRESSION$E))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Outlier analysis
Clustering of initial normalised data (with NULL design)
```{r decompse.normalise.data, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'ID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=CDR, shape=SEX, size=PMI))
p <- p + theme_bw() + theme(legend.position="top") + scale_shape_manual(values = 1:7)
p <- p + geom_text(aes(label= ID), size=4, hjust=1)
p

COVARIATES.tmp = data.matrix(COVARIATES)
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(VOOM.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);
tmp = gsub('\\..*','',cols)
WGCNA::plotDendroAndColors(tree, 
                           colors = tmp, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
dev.off()
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, cache=TRUE}
# Initial design matrix with covariates under consideration
DM = getDesignMatrix(COVARIATES, Intercept = F)

# Find correlation between PC's of gene expression with covariates
DESIGN = getDesignMat2Fact(DM$design,FactorCovariates)
rownames(DESIGN) = rownames(COVARIATES)
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, DESIGN,'NULL design(voom-normalized)', isKeyPlot=TRUE)

# Find significant covariates
adjustCovars = designMatVarsToCovars(getCovariatesMap(DM), preAdjustedSigCovars$significantCovars)
```
Significant covariates to adjust at FDR 0.1 are `r adjustCovars`
```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative design)
Since many covariates are correlated, re-normalising COVARIATES with an iterative design matrix. Here BB(Donor Information) is chosen as random effect
```{r iterativeNorm}
# Primary variable of interest
postAdjustCovars = 'CDR'

# Assign residual covariates
residualSigCovars = preAdjustedSigCovars

significantCovars = designMatVarsToCovars(getCovariatesMap(DM), 
                                          residualSigCovars$significantCovars)

covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),
                                         residualSigCovars$Effects.significantCovars)

covariatesEffects = covariatesEffects[setdiff(significantCovars, postAdjustCovars)]

postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 100){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
  
  # Estimate correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F)
  
  # Fit a linear model
  FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), postAdjustCovars)
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  tmp = sapply(dplyr::select(DESIGN,-one_of(colnames(DM1$design))), as.numeric)
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                tmp, 'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  significantCovars = designMatVarsToCovars(getCovariatesMap(DM), residualSigCovars$significantCovars)
  covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),residualSigCovars$Effects.significantCovars)
  covariatesEffects = covariatesEffects[setdiff(significantCovars, postAdjustCovars)]
  
  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`

Degenerate variables are:
```{r degenVar}
# Post adjusted design matrix
DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F], Intercept = F)
tmp = linColumnFinder(DM1$design)
print(tmp$relations)
DM1$design = DM1$design[,tmp$indepCols] 

# Recalculate weights with new design
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, 
                                     design=DM1$design, plot=F)

# Linear model with new weights and design
VOOM.ADJUSTED.FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION)  

# Residuals after normalisation
RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
rownames(DESIGN) <- rownames(COVARIATES)
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION,
                                              DESIGN,
                                              'all adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
```

```{r residualSigCovars.manual, fig.width=20, fig.height=12}
residualSigCovars[["PC_res"]][[2]]$plotData
```
### Residual calculation
Calculate weighted residuals and add back "CDR" to the residuals
```{r varsToAddBack}
# Add variable of interest back to the residuals
varsToAddIn = c("CDR0","CDR0.5","CDR1","CDR2","CDR3","CDR4","CDR5")
RESIDUAL.GENE_EXPRESSION = RESIDUAL.GENE_EXPRESSION + 
  VOOM.ADJUSTED.FIT$coefficients[,varsToAddIn] %*% t(DM1$design[,varsToAddIn])
```
Coexpression of genes
```{r coexp2}
cr = cor(t(RESIDUAL.GENE_EXPRESSION))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```
### Clustering residual data
```{r decompse.normalise.data2, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'ID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=factor(Batch), shape=as.factor(CDR), size=as.numeric(PMI)))
p <- p + theme_bw() + theme(legend.position="top") + scale_shape_manual(values = 1:7)
p <- p + geom_text(aes(label= ID), size=4, hjust=1)
p

COVARIATES.tmp = data.matrix(COVARIATES)
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(RESIDUAL.GENE_EXPRESSION)))
cols = WGCNA::labels2colors(COVARIATES.tmp);
tmp = gsub('\\..*','',cols)
WGCNA::plotDendroAndColors(tree, 
                           colors = tmp, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```
```{r get.gene.symbols, cache=TRUE}
# Define biomart object
mart <- useMart(host = "jul2016.archive.ensembl.org", dataset = "hsapiens_gene_ensembl", biomart = "ENSEMBL_MART_ENSEMBL")

RESIDUAL.GENE_EXPRESSION = rownameToFirstColumn(RESIDUAL.GENE_EXPRESSION, 'GeneID') %>%
  dplyr::mutate(gene.id = GeneID) %>%
  tidyr::separate(GeneID, c('ensembl_gene_id','position'),sep = '\\.')

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                       filters = "ensembl_gene_id", values = RESIDUAL.GENE_EXPRESSION$ensembl_gene_id,
                       mart = mart)
RESIDUAL.GENE_EXPRESSION <- left_join(RESIDUAL.GENE_EXPRESSION, Ensemble2HGNC) %>%
  dplyr::select(gene.id, ensembl_gene_id, hgnc_symbol, one_of(rownames(COVARIATES)))
```

```{r summary, results='asis', echo=FALSE}
writeLines('Number of samples in each category are')
tmp = as.data.frame(summary(COVARIATES$CDR))
colnames(tmp) = 'No. of samples'
kable(tmp)
```
### Differential expression analysis
Genes that are differentially expressed at an FDR <= 0.05 are
```{r diffExp, fig.height=10, fig.width=18}
# Fit contrast
contrast = colnames(VOOM.ADJUSTED.FIT$coefficients) %>%
  grep('CDR',.,value=T) %>%
  combn(2) %>%
  apply(2,paste,collapse='-') %>%
  makeContrasts(contrasts=.,
                levels = colnames(VOOM.ADJUSTED.FIT$coefficients))
FIT = contrasts.fit(VOOM.ADJUSTED.FIT, contrasts=contrast)
FIT = eBayes(FIT)

# Get differnetial expression
DE = list()
p = list()
for (i in 1:dim(contrast)[2]){
  DE[[i]] = topTable(FIT, coef=i, number = dim(VOOM.ADJUSTED.GENE_EXPRESSION$E)[1])
  DE[[i]] = rownameToFirstColumn(DE[[i]], "gene.id") %>%
    dplyr::mutate(GeneID = gene.id) %>%
    tidyr::separate(GeneID, c('ensembl_gene_id','position'),sep = '\\.') %>%
    left_join(Ensemble2HGNC)
  p[[i]] <- ggplot(DE[[i]], aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()
  p[[i]] <- p[[i]] + ggtitle(gsub('TISSUE.CDR','',colnames(contrast)[i])) + xlim(c(-1.5,1.5)) + ylim(c(0,6))
  p[[i]] <- p[[i]] + geom_hline(yintercept = -log10(0.05), color = 'red')
}
# multiplot(plotlist=p, cols = 3)
names(DE) = gsub('CDR','',colnames(contrast))

DE = DE %>% 
  rbindlist(idcol = 'Comparison', use.names=T, fill=T)

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(Comparison, hgnc_symbol) %>%
  group_by(Comparison) %>%
  dplyr::summarise(nDiffExp = length(unique(hgnc_symbol)), 
                   DiffExpGenes = paste(unique(hgnc_symbol), collapse = ','))
kable(tmp)
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Code
CODE <- Folder(name = "MSBB BM 36 RNASeq  RealignedFromSource StarHTSeq", parentId = parentId)
CODE <- synStore(CODE)

# Expression
write.table(RESIDUAL.GENE_EXPRESSION, file='expression.tsv', sep = '\t', row.names=F, quote=F)
obj = File('expression.tsv', 
           name = 'Voom Adjusted Weighted Residual Expression (CDR added)', 
           parentId = CODE$properties$id)
expr_obj = synStore(obj,
                    used = ALL_USED_IDs, 
                    activityName = activityName, 
                    executed = thisFile, 
                    activityDescription = activityDescription)

# Covariates
covar = rownameToFirstColumn(COVARIATES, 'SampleId')
write.table(covar, file='covariates.tsv', sep = '\t', row.names=F, quote=F)
obj = File('covariates.tsv', 
           name = 'Covariates', 
           parentId = CODE$properties$id)
covar_obj = synStore(obj,
                     used = ALL_USED_IDs, 
                     activityName = activityName, 
                     executed = thisFile, 
                     activityDescription = activityDescription)

# Store differential expression results
write.table(DE, file = 'diffExpression.tsv', sep = '\t', row.names=F, quote=F)
dexp_obj = File('diffExpression.tsv', name = 'Differential Expression Analysis', parentId = CODE$properties$id)
dexp_obj = synStore(dexp_obj, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
```
|  *Results*                |  *SynapseID*                    |
|  -------                  |   ---------                     |
|  Covariates               |  `r covar_obj$properties$id`    |
|  Residual Expression      |  `r expr_obj$properties$id`     |
|  Differential Expression  |  `r dexp_obj$properties$id`     |